{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "774a56de",
   "metadata": {},
   "source": [
    "# Estimate Protein Isoelectric Point\n",
    "Lee Yuan Zhe  \n",
    "31/12/2021"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ea356e8",
   "metadata": {},
   "source": [
    "## 1. Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2dedc01c",
   "metadata": {},
   "source": [
    "The isoelectric point of a protein, pI for short, is the pH at which the protein has a net electrical charge of zero. When the pH of the environment is above the pI of a protein, the protein will have a negative net charge. The converse is true when the pH of the environment is below the pI. Hence, knowing the pI of a protein allows us to control the charge of the protein by varying the pH of the environment. With the application of suitable external electric field, it is possible to isolate specific proteins of interest for further experimentation.\n",
    "\n",
    "In this project, we aim to estimate the pI of various proteins using simple algorithms. Roughly speaking, the algorithms do so by computing the net charge of a protein at various pH values until they find one pH value which gives zero net charge. To gauge the accuracy of our estimation, we compare our calculated pI values with the values available from the [Proteome Isoelecric Point Database](http://isoelectricpointdb2.mimuw.edu.pl/index.html). With coding, we can test our algorithms quickly, allowing for prompt amendments when necessary.\n",
    "\n",
    "The following modules will be used in this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c42a4f41",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e039a2c",
   "metadata": {},
   "source": [
    "## 2. pK$_a$ Values of Various Amino Acids"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78d140cd",
   "metadata": {},
   "source": [
    "### 2.1 Amino Acids"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe500e21",
   "metadata": {},
   "source": [
    "Each protein is made up of a chain of amino acids. The structure of a neutral amino acid is shown in the figure below.\n",
    "<img src=\"amino_acid.png\" width=\"200\"/>\n",
    "\n",
    "Each amino acid has two terminals: the NH$_2$- group forms the N-terminal while the -COOH group forms the C-terminal. The R group is the side chain specific to each amino acid. Two amino acids are linked together when the C-terminal of one amino acid loses an -OH group while the N-terminal of the other amino acid loses an -H to form an amide linkage, represented by -CONH as shown in the figure below.\n",
    "<img src='amide_linkage.png' width='600'/>\n",
    "<center><em>The elements in blue are lost by each amino acid during the formation of an amide linkage, shown in red.</em></center>\n",
    "\n",
    "A protein is formed when many amino acids are linked to each other through amide linkages. Thus, every protein will always have one N-terminal, one C-terminal, and various R groups depending on the amino acids used to from the protein. The N-terminal is basic. It can gain a proton, H$^+$, giving the protein a positive charge. On the other hand, the C-terminal is acidic, which allows it to lose a proton, giving the protein a negative charge. If the R group is basic or acidic, it can also contribute to the protein charge. There are also neutral R groups which will not contribute to the charge of the protein. \n",
    "\n",
    "Hence, the overall charge of a protein at a pH is dependent on the charge of its N-terminal, C-terminal, and the various R-groups at that pH."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c10590a",
   "metadata": {},
   "source": [
    "### 2.2 Acid Dissociation Constant, K$_a$"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30d805be",
   "metadata": {},
   "source": [
    "The strength of an acid is measured by its K$_a$ value, which is a constant. For an acid undergoing the following dissociation in water,\n",
    "\\begin{align}\n",
    "\\text{HA} \\rightleftharpoons \\text{H}^+ + \\text{A}^-\n",
    "\\end{align}\n",
    "The mathematical definition of K$_a$ is given by:\n",
    "\\begin{equation}\n",
    "K_a = \\frac{[\\text{H}^+][\\text{A}^-]}{[\\text{HA}]}\n",
    "\\tag{1}\n",
    "\\end{equation}\n",
    "\n",
    "For basic groups, the K$_a$ value gives the strength of the conjugate acid, which is the protonated base. One such example is ammonia as shown below.\n",
    "\\begin{align}\n",
    "\\text{NH}_4^+ \\rightleftharpoons \\text{H}^+ + \\text{NH}_3\n",
    "\\end{align}\n",
    "\\begin{align}\n",
    "K_a = \\frac{[\\text{H}^+][\\text{NH}_3]}{[\\text{NH}_4^+]}\n",
    "\\tag{2}\n",
    "\\end{align}\n",
    "\n",
    "From the equations, we can infer that the concentration of dissociated acids and protonated bases depends on the respective K$_a$ values and the concentration of protons in the solution. Hence, in order to compute the charge of a protein, we need all the K$_a$ values of all basic and acid groups of every amino acid group making up the protein. We will also need the nature of each R group as this will determine whether it possesses a negative or positive charge.\n",
    "\n",
    "The values are obtained from [Sigma Aldrich](https://www.sigmaaldrich.com/SG/en/technical-documents/technical-article/protein-biology/protein-structural-analysis/amino-acid-reference-chart). The chart lists the pK$_a$ values for common amino acids at 25$^{\\circ}$C, where pK$_a$ = $-\\log$ K$_a$. 'pKa1', 'pKa2', and 'pKa3' give the pK$_a$ values of the C-terminal, N-terminal, and R group respectively."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "acade834",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Abbr</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Molecular Weight</th>\n",
       "      <th>Molecular Formula</th>\n",
       "      <th>Residue Formula</th>\n",
       "      <th>Residue Weight (-H2O)</th>\n",
       "      <th>pKa1</th>\n",
       "      <th>pKa2</th>\n",
       "      <th>pKa3</th>\n",
       "      <th>pI</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Alanine</th>\n",
       "      <td>Ala</td>\n",
       "      <td>A</td>\n",
       "      <td>89.10</td>\n",
       "      <td>C3H7NO2</td>\n",
       "      <td>C3H5NO</td>\n",
       "      <td>71.08</td>\n",
       "      <td>2.34</td>\n",
       "      <td>9.69</td>\n",
       "      <td>-</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Arginine</th>\n",
       "      <td>Arg</td>\n",
       "      <td>R</td>\n",
       "      <td>174.20</td>\n",
       "      <td>C6H14N4O2</td>\n",
       "      <td>C6H12N4O</td>\n",
       "      <td>156.19</td>\n",
       "      <td>2.17</td>\n",
       "      <td>9.04</td>\n",
       "      <td>12.48</td>\n",
       "      <td>10.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Asparagine</th>\n",
       "      <td>Asn</td>\n",
       "      <td>N</td>\n",
       "      <td>132.12</td>\n",
       "      <td>C4H8N2O3</td>\n",
       "      <td>C4H6N2O2</td>\n",
       "      <td>114.11</td>\n",
       "      <td>2.02</td>\n",
       "      <td>8.8</td>\n",
       "      <td>-</td>\n",
       "      <td>5.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Aspartic acid</th>\n",
       "      <td>Asp</td>\n",
       "      <td>D</td>\n",
       "      <td>133.11</td>\n",
       "      <td>C4H7NO4</td>\n",
       "      <td>C4H5NO3</td>\n",
       "      <td>115.09</td>\n",
       "      <td>1.88</td>\n",
       "      <td>9.6</td>\n",
       "      <td>3.65</td>\n",
       "      <td>2.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cysteine</th>\n",
       "      <td>Cys</td>\n",
       "      <td>C</td>\n",
       "      <td>121.16</td>\n",
       "      <td>C3H7NO2S</td>\n",
       "      <td>C3H5NOS</td>\n",
       "      <td>103.15</td>\n",
       "      <td>1.96</td>\n",
       "      <td>10.28</td>\n",
       "      <td>8.18</td>\n",
       "      <td>5.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Glutamic acid</th>\n",
       "      <td>Glu</td>\n",
       "      <td>E</td>\n",
       "      <td>147.13</td>\n",
       "      <td>C5H9NO4</td>\n",
       "      <td>C5H7NO3</td>\n",
       "      <td>129.12</td>\n",
       "      <td>2.19</td>\n",
       "      <td>9.67</td>\n",
       "      <td>4.25</td>\n",
       "      <td>3.22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Glutamine</th>\n",
       "      <td>Gln</td>\n",
       "      <td>Q</td>\n",
       "      <td>146.15</td>\n",
       "      <td>C5H10N2O3</td>\n",
       "      <td>C5H8N2O2</td>\n",
       "      <td>128.13</td>\n",
       "      <td>2.17</td>\n",
       "      <td>9.13</td>\n",
       "      <td>-</td>\n",
       "      <td>5.65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Glycine</th>\n",
       "      <td>Gly</td>\n",
       "      <td>G</td>\n",
       "      <td>75.07</td>\n",
       "      <td>C2H5NO2</td>\n",
       "      <td>C2H3NO</td>\n",
       "      <td>57.05</td>\n",
       "      <td>2.34</td>\n",
       "      <td>9.6</td>\n",
       "      <td>-</td>\n",
       "      <td>5.97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Histidine</th>\n",
       "      <td>His</td>\n",
       "      <td>H</td>\n",
       "      <td>155.16</td>\n",
       "      <td>C6H9N3O2</td>\n",
       "      <td>C6H7N3O</td>\n",
       "      <td>137.14</td>\n",
       "      <td>1.82</td>\n",
       "      <td>9.17</td>\n",
       "      <td>6</td>\n",
       "      <td>7.59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hydroxyproline</th>\n",
       "      <td>Hyp</td>\n",
       "      <td>O</td>\n",
       "      <td>131.13</td>\n",
       "      <td>C5H9NO3</td>\n",
       "      <td>C5H7NO2</td>\n",
       "      <td>113.11</td>\n",
       "      <td>1.82</td>\n",
       "      <td>9.65</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Isoleucine</th>\n",
       "      <td>Ile</td>\n",
       "      <td>I</td>\n",
       "      <td>131.18</td>\n",
       "      <td>C6H13NO2</td>\n",
       "      <td>C6H11NO</td>\n",
       "      <td>113.16</td>\n",
       "      <td>2.36</td>\n",
       "      <td>9.6</td>\n",
       "      <td>-</td>\n",
       "      <td>6.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Leucine</th>\n",
       "      <td>Leu</td>\n",
       "      <td>L</td>\n",
       "      <td>131.18</td>\n",
       "      <td>C6H13NO2</td>\n",
       "      <td>C6H11NO</td>\n",
       "      <td>113.16</td>\n",
       "      <td>2.36</td>\n",
       "      <td>9.6</td>\n",
       "      <td>-</td>\n",
       "      <td>5.98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lysine</th>\n",
       "      <td>Lys</td>\n",
       "      <td>K</td>\n",
       "      <td>146.19</td>\n",
       "      <td>C6H14N2O2</td>\n",
       "      <td>C6H12N2O</td>\n",
       "      <td>128.18</td>\n",
       "      <td>2.18</td>\n",
       "      <td>8.95</td>\n",
       "      <td>10.53</td>\n",
       "      <td>9.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Methionine</th>\n",
       "      <td>Met</td>\n",
       "      <td>M</td>\n",
       "      <td>149.21</td>\n",
       "      <td>C5H11NO2S</td>\n",
       "      <td>C5H9NOS</td>\n",
       "      <td>131.20</td>\n",
       "      <td>2.28</td>\n",
       "      <td>9.21</td>\n",
       "      <td>-</td>\n",
       "      <td>5.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Phenylalanine</th>\n",
       "      <td>Phe</td>\n",
       "      <td>F</td>\n",
       "      <td>165.19</td>\n",
       "      <td>C9H11NO2</td>\n",
       "      <td>C9H9NO</td>\n",
       "      <td>147.18</td>\n",
       "      <td>1.83</td>\n",
       "      <td>9.13</td>\n",
       "      <td>-</td>\n",
       "      <td>5.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Proline</th>\n",
       "      <td>Pro</td>\n",
       "      <td>P</td>\n",
       "      <td>115.13</td>\n",
       "      <td>C5H9NO2</td>\n",
       "      <td>C5H7NO</td>\n",
       "      <td>97.12</td>\n",
       "      <td>1.99</td>\n",
       "      <td>10.6</td>\n",
       "      <td>-</td>\n",
       "      <td>6.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pyroglutamatic</th>\n",
       "      <td>Glp</td>\n",
       "      <td>U</td>\n",
       "      <td>139.11</td>\n",
       "      <td>C5H7NO3</td>\n",
       "      <td>C5H5NO2</td>\n",
       "      <td>121.09</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>5.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Serine</th>\n",
       "      <td>Ser</td>\n",
       "      <td>S</td>\n",
       "      <td>105.09</td>\n",
       "      <td>C3H7NO3</td>\n",
       "      <td>C3H5NO2</td>\n",
       "      <td>87.08</td>\n",
       "      <td>2.21</td>\n",
       "      <td>9.15</td>\n",
       "      <td>-</td>\n",
       "      <td>5.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Threonine</th>\n",
       "      <td>Thr</td>\n",
       "      <td>T</td>\n",
       "      <td>119.12</td>\n",
       "      <td>C4H9NO3</td>\n",
       "      <td>C4H7NO2</td>\n",
       "      <td>101.11</td>\n",
       "      <td>2.09</td>\n",
       "      <td>9.1</td>\n",
       "      <td>-</td>\n",
       "      <td>5.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tryptophan</th>\n",
       "      <td>Trp</td>\n",
       "      <td>W</td>\n",
       "      <td>204.23</td>\n",
       "      <td>C11H12N2O2</td>\n",
       "      <td>C11H10N2O</td>\n",
       "      <td>186.22</td>\n",
       "      <td>2.83</td>\n",
       "      <td>9.39</td>\n",
       "      <td>-</td>\n",
       "      <td>5.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tyrosine</th>\n",
       "      <td>Tyr</td>\n",
       "      <td>Y</td>\n",
       "      <td>181.19</td>\n",
       "      <td>C9H11NO3</td>\n",
       "      <td>C9H9NO2</td>\n",
       "      <td>163.18</td>\n",
       "      <td>2.2</td>\n",
       "      <td>9.11</td>\n",
       "      <td>10.07</td>\n",
       "      <td>5.66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Valine</th>\n",
       "      <td>Val</td>\n",
       "      <td>V</td>\n",
       "      <td>117.15</td>\n",
       "      <td>C5H11NO2</td>\n",
       "      <td>C5H9NO</td>\n",
       "      <td>99.13</td>\n",
       "      <td>2.32</td>\n",
       "      <td>9.62</td>\n",
       "      <td>-</td>\n",
       "      <td>5.96</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Abbr Symbol  Molecular Weight Molecular Formula  \\\n",
       "Name                                                             \n",
       "Alanine         Ala      A             89.10           C3H7NO2   \n",
       "Arginine        Arg      R            174.20         C6H14N4O2   \n",
       "Asparagine      Asn      N            132.12          C4H8N2O3   \n",
       "Aspartic acid   Asp      D            133.11           C4H7NO4   \n",
       "Cysteine        Cys      C            121.16          C3H7NO2S   \n",
       "Glutamic acid   Glu      E            147.13           C5H9NO4   \n",
       "Glutamine       Gln      Q            146.15         C5H10N2O3   \n",
       "Glycine         Gly      G             75.07           C2H5NO2   \n",
       "Histidine       His      H            155.16          C6H9N3O2   \n",
       "Hydroxyproline  Hyp      O            131.13           C5H9NO3   \n",
       "Isoleucine      Ile      I            131.18          C6H13NO2   \n",
       "Leucine         Leu      L            131.18          C6H13NO2   \n",
       "Lysine          Lys      K            146.19         C6H14N2O2   \n",
       "Methionine      Met      M            149.21         C5H11NO2S   \n",
       "Phenylalanine   Phe      F            165.19          C9H11NO2   \n",
       "Proline         Pro      P            115.13           C5H9NO2   \n",
       "Pyroglutamatic  Glp      U            139.11           C5H7NO3   \n",
       "Serine          Ser      S            105.09           C3H7NO3   \n",
       "Threonine       Thr      T            119.12           C4H9NO3   \n",
       "Tryptophan      Trp      W            204.23        C11H12N2O2   \n",
       "Tyrosine        Tyr      Y            181.19          C9H11NO3   \n",
       "Valine          Val      V            117.15          C5H11NO2   \n",
       "\n",
       "               Residue Formula  Residue Weight (-H2O)  pKa1   pKa2   pKa3  \\\n",
       "Name                                                                        \n",
       "Alanine                 C3H5NO                  71.08  2.34   9.69      -   \n",
       "Arginine              C6H12N4O                 156.19  2.17   9.04  12.48   \n",
       "Asparagine            C4H6N2O2                 114.11  2.02    8.8      -   \n",
       "Aspartic acid          C4H5NO3                 115.09  1.88    9.6   3.65   \n",
       "Cysteine               C3H5NOS                 103.15  1.96  10.28   8.18   \n",
       "Glutamic acid          C5H7NO3                 129.12  2.19   9.67   4.25   \n",
       "Glutamine             C5H8N2O2                 128.13  2.17   9.13      -   \n",
       "Glycine                 C2H3NO                  57.05  2.34    9.6      -   \n",
       "Histidine              C6H7N3O                 137.14  1.82   9.17      6   \n",
       "Hydroxyproline         C5H7NO2                 113.11  1.82   9.65      -   \n",
       "Isoleucine             C6H11NO                 113.16  2.36    9.6      -   \n",
       "Leucine                C6H11NO                 113.16  2.36    9.6      -   \n",
       "Lysine                C6H12N2O                 128.18  2.18   8.95  10.53   \n",
       "Methionine             C5H9NOS                 131.20  2.28   9.21      -   \n",
       "Phenylalanine           C9H9NO                 147.18  1.83   9.13      -   \n",
       "Proline                 C5H7NO                  97.12  1.99   10.6      -   \n",
       "Pyroglutamatic         C5H5NO2                 121.09     -      -      -   \n",
       "Serine                 C3H5NO2                  87.08  2.21   9.15      -   \n",
       "Threonine              C4H7NO2                 101.11  2.09    9.1      -   \n",
       "Tryptophan           C11H10N2O                 186.22  2.83   9.39      -   \n",
       "Tyrosine               C9H9NO2                 163.18   2.2   9.11  10.07   \n",
       "Valine                  C5H9NO                  99.13  2.32   9.62      -   \n",
       "\n",
       "                   pI  \n",
       "Name                   \n",
       "Alanine             6  \n",
       "Arginine        10.76  \n",
       "Asparagine       5.41  \n",
       "Aspartic acid    2.77  \n",
       "Cysteine         5.07  \n",
       "Glutamic acid    3.22  \n",
       "Glutamine        5.65  \n",
       "Glycine          5.97  \n",
       "Histidine        7.59  \n",
       "Hydroxyproline      -  \n",
       "Isoleucine       6.02  \n",
       "Leucine          5.98  \n",
       "Lysine           9.74  \n",
       "Methionine       5.74  \n",
       "Phenylalanine    5.48  \n",
       "Proline           6.3  \n",
       "Pyroglutamatic   5.68  \n",
       "Serine           5.68  \n",
       "Threonine         5.6  \n",
       "Tryptophan       5.89  \n",
       "Tyrosine         5.66  \n",
       "Valine           5.96  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import csv file as data frame.\n",
    "amino_acid=pd.read_csv('amino_acids_data.csv',index_col='Name')\n",
    "amino_acid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f29eb10e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Abbr                      object\n",
       "Symbol                    object\n",
       "Molecular Weight         float64\n",
       "Molecular Formula         object\n",
       "Residue Formula           object\n",
       "Residue Weight (-H2O)    float64\n",
       "pKa1                      object\n",
       "pKa2                      object\n",
       "pKa3                      object\n",
       "pI                        object\n",
       "dtype: object"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check the data type to confirm whether the entries can be used for calculations later.\n",
    "amino_acid.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "08659c28",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Replace all missing values with NaN.\n",
    "amino_acid.replace('-', np.nan, inplace = True)\n",
    "\n",
    "# Create a dictionary for astype.\n",
    "data_type = dict(zip(amino_acid.columns.to_list()[-4:], ['float64']*4))\n",
    "\n",
    "# Change all entries for the last 4 columns to float.\n",
    "amino_acid = amino_acid.astype(data_type)\n",
    "\n",
    "# Create a dictionary matching the pKa values of each N-terminal and C-terminal to its respective amino acid.\n",
    "pKaDict_NH2 = dict(zip(amino_acid['Symbol'], amino_acid['pKa2']))\n",
    "pKaDict_COOH = dict(zip(amino_acid['Symbol'], amino_acid['pKa1']))\n",
    "\n",
    "# Select only R groups with pKa values.\n",
    "R = amino_acid[amino_acid['pKa3'] > 0][['Symbol', 'pKa3']]\n",
    "\n",
    "# Create a dictionary matching the pKa values of each R group to its respective amino acid.\n",
    "pKaDict_R = dict(zip(R['Symbol'], R['pKa3']))\n",
    "\n",
    "# Create a dictionary identifying the nature of the R group of each amino acid: True if acidic, False if basic.\n",
    "isAcid = {'K':False, 'R':False, 'H':False, 'E':True, 'D':True, 'Y':True, 'C':True}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e7bab28",
   "metadata": {},
   "source": [
    "## 3. Estimating Protein Isoelectric Point"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c37b2395",
   "metadata": {},
   "source": [
    "### 3.1 Estimate The Charge of A Protein at Different pH Values"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e91940b",
   "metadata": {},
   "source": [
    "To estimate the charge of a protein, we compute the charge of each R group, N-terminal and C-terminal separately. The final charge of the protein will be the sum of the charges of all groups. The magnitude of the charge is computed from the ratio of the concentration of the dissociated species to the total concentration of the species before dissociation, $\\frac{[\\text{A}^-]}{[\\text{HA}] + [\\text{A}^-]}$. For an acid, from equation (1):\n",
    "\\begin{align}\n",
    "K_a &= \\frac{[\\text{H}^+][\\text{A}^-]}{[\\text{HA}]} \\\\\n",
    "-\\log K_a &= -\\log [\\text{H}^+] - \\log \\frac{[\\text{A}^-]}{[\\text{HA}]} \\\\\n",
    "\\text{p}K_a &= \\text{pH} - \\log \\frac{[\\text{A}^-]}{[\\text{HA}]} \\\\\n",
    "\\log \\frac{[\\text{A}^-]}{[\\text{HA}]} &= \\text{pH} - \\text{p}K_a \\\\\n",
    "\\frac{[\\text{A}^-]}{[\\text{HA}]} &= 10^{\\text{pH} - \\text{p}K_a} \\\\\n",
    "&= r \\\\\n",
    "\\text{dissociated ratio} &= \\frac{r}{r+1} \\\\\n",
    "\\text{charge} &= - \\text{ dissociated ratio}\n",
    "\\end{align}\n",
    "\n",
    "For a base, we can use the exact steps to arrive at the same expression for the dissociated ratio. However, as seen from equation (2), the charge does not come from the dissociated species but the species before dissociation. The charge is also of opposite sign. Hence,\n",
    "\\begin{align}\n",
    "\\text{charge} &= \\frac{[\\text{NH}_4^+]}{[\\text{NH}_4^+]+[\\text{NH}_3]} \\\\\n",
    "&= 1 - \\text{dissociated ratio}\n",
    "\\end{align}\n",
    "\n",
    "The mathematical derivations form the basis of the estimateCharge function defined below. In all our following calculations, we assume that the proteins are in water of 25$^{\\circ}$C. We also ignore the effect of neighbouring amino acids on the dissociation constants of each component amino acid. In reality, the pK$_a$ values depend on how the component amino acids interact with each other."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0db6bf7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def estimateCharge(sequence, pH):\n",
    "    \"\"\"Use pKa values to estimate the charge of a sequence of amino acids at a given pH\"\"\"\n",
    "    \n",
    "    # Starting value for the net charge of protein.\n",
    "    total = 0.0\n",
    "    \n",
    "    # By convention, the first letter of an amino acid sequence is the N-terminal. We begin by obtainning the pKa value of the \n",
    "    # N-terminal of the first amino acid by using the dictionary built earlier. The charge is computed from the pKa value using \n",
    "    # the derivation shown earlier.\n",
    "    pKaN = pKaDict_NH2.get(sequence[0])\n",
    "    rN = 10.0 ** (pH - pKaN)\n",
    "    dissociatedN = rN / (rN + 1.0)\n",
    "    chargeN = 1.0 - dissociatedN\n",
    "    \n",
    "    # Update the net charge of protein with the charge from the N-terminal.\n",
    "    total += chargeN\n",
    "    \n",
    "    # Loop over the entire sequence to compute the charge due to each R group.\n",
    "    for aminoAcid in sequence:\n",
    "        \n",
    "        # Obtain the pKa value of the R group from the dictionary built earlier.\n",
    "        pKaR = pKaDict_R.get(aminoAcid)\n",
    "        \n",
    "        # Check whether the R group is neutral. If it is not, calculate the dissociated ratio.\n",
    "        if pKaR is not None:\n",
    "            rR = 10.0 ** (pH - pKaR)\n",
    "            dissociatedR = rR / (rR + 1.0)\n",
    "            \n",
    "            # Check whether the R group is acidic or basic. Calculate the charge accordingly.\n",
    "            if isAcid[aminoAcid]:\n",
    "                chargeR = -1.0 * dissociatedR\n",
    "            else:\n",
    "                chargeR = 1.0 - dissociatedR\n",
    "            \n",
    "            # Update the net charge of protein with the charge of the R group.\n",
    "            total += chargeR\n",
    "    \n",
    "    # Obtain the pKa value of the C-terminal using the dictionary built earlier, followed by charge calculation.\n",
    "    pKaC = pKaDict_COOH.get(sequence[-1])\n",
    "    rC = 10.0 ** (pH - pKaC)\n",
    "    dissociatedC = rC / (rC + 1.0)\n",
    "    chargeC = -1.0 * dissociatedC\n",
    "    \n",
    "    # Update the net charge of protein with the charge of the C-terminal.\n",
    "    total += chargeC\n",
    "    return total"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e63e9cd1",
   "metadata": {},
   "source": [
    "With the estimateCharge function, we can compute the charge of the protein at every pH value. By running the function through a loop where the pH value of each iteration differs from the next by a small amount, we can find the pH value which gives the charge which is closest to zero. This gives us the estimated pI value of the protein. However, this exhaustive method takes too many steps. Hence, we will employ an optimisation algorithm to find the pI value."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad636a94",
   "metadata": {},
   "source": [
    "### 3.2 Optimisation Algorithm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "771d9398",
   "metadata": {},
   "source": [
    "To come up with an effective optimisation algorithm, we analyse how the charge of an example sequence, ARN, varies with pH."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6c11b5b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAksAAAHFCAYAAADi7703AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAA9hAAAPYQGoP6dpAABq2UlEQVR4nO3dd3hT1f8H8PdN0qS7tJROulhllVVWGQKyZAkOUPHLUpGhICCKFRVxUGUoKEuGoCIIyBBBEGTJKHtDKau0BVpaKLSlK+v8/qjNj9JJaXub9P16njw0N+fmvnObJh/OvfccSQghQERERET5UsgdgIiIiKgiY7FEREREVAgWS0RERESFYLFEREREVAgWS0RERESFYLFEREREVAgWS0RERESFYLFEREREVAgWS0RERESFYLFEpeLMmTMYNmwYAgICYG1tDXt7ezRr1gzTp09HUlKSqZ2/vz969+4tY1J57dy5E82bN4ednR0kScLGjRsLbBsbG4vRo0ejTp06sLGxgYuLC4KCgjB8+HDExsaWX2gLptPp4OHhAUmS8Pvvv+fb5tNPP4UkSaablZUVfH19MXz4cMTHx+dp7+/vD0mSMHLkyDyP7dmzp9BtVURDhw6Fvb19gY/b29tj6NChpbY9rVaLkSNHwtPTE0qlEk2aNCmwrRACv/32G9q3bw83NzdYW1ujevXq6N69O5YsWVKi7fv7+5f49axcuRKzZ8/O9zFJkvDpp5+W6HlJfiq5A5D5W7x4MUaPHo3AwEC89957qF+/PnQ6HY4dO4aFCxciPDwcGzZskDum7IQQGDBgAOrUqYNNmzbBzs4OgYGB+ba9ceMGmjVrhipVquDdd99FYGAgkpOTceHCBaxZswbXrl2Dj49POb8Cy7N582bcvn0bALB06VK8+OKLBbbdtm0bnJyc8ODBA2zfvh2zZs3CwYMHcerUKVhZWeVpv3TpUowfP77A3zHlb8GCBfjhhx/w/fffIzg4uNBCLTQ0FF9//TWGDx+O9957Dw4ODoiOjsauXbvwxx9/4I033ijH5NnF0rlz5zBu3Lg8j4WHh6N69erlmodKD4sleiLh4eEYNWoUunbtio0bN0Kj0Zge69q1K959911s27atXDMZDAbo9fpcWSqCW7duISkpCc899xw6d+5caNvFixfjzp07OHLkCAICAkzL+/Xrhw8//BBGo7Gs41YKS5cuhVqtRocOHbB9+3bcuHGjwC+04OBguLq6AgC6dOmCO3fuYNmyZdi/fz86deqUq21ISAguXLiADz/8EOvWrSvz12FJzp07BxsbG7z99tuFtsvIyMDs2bMxePBgLFq0KNdjQ4cOrXB/I61bt5Y7Aj0BHoajJzJt2jRIkoRFixblW5yo1Wo8++yzeZZv27YNzZo1g42NDerWrYsff/wx1+OJiYkYPXo06tevD3t7e7i5ueHpp5/Gvn37crW7fv06JEnC9OnT8cUXXyAgIAAajQa7d+8GAPzxxx9o1KgRNBoNatSogTlz5pgOqzxMCIH58+ejSZMmsLGxgbOzM1588UVcu3atWPth//796Ny5MxwcHGBra4s2bdpgy5Ytpsc//fRT05fwpEmTIEkS/P39C3y+u3fvQqFQwM3NLd/HFYrcf7rHjh3Ds88+CxcXF1hbW6Np06ZYs2ZNnvUOHTqEtm3bwtraGl5eXggNDcXixYshSRKuX79ualfQIYP8DlHEx8djxIgRqF69OtRqNQICAjB16lTo9XpTm5zf08yZM/HNN98gICAA9vb2CAkJwaFDh/Js5/Dhw+jTpw+qVq0Ka2tr1KxZM8//1i9fvoyBAwfCzc0NGo0G9erVw7x58/LdX/m5desWtm3bhj59+uC9996D0WjE8uXLi71+8+bNAcDUM/UwFxcXfPDBB1i/fn2+r68wiYmJUKvV+Pjjj/M8dvHiRUiShO+++w4AkJ6ejokTJ5oOf7u4uKB58+ZYtWpVodtYvnw5JEnCjh07MGzYMLi4uMDOzg59+vQp9nv+cWVmZiI0NBQBAQFQq9Xw9vbGW2+9hfv375vaSJKEJUuWICMjw3TYs6DfSVpaGrKysuDp6Znv44/+jSQlJWH06NHw9vaGWq1GjRo1MHnyZGRlZRWaO2dfPfz3Afz/IdU9e/YAADp27IgtW7YgOjo612Hbh1/bo39T586dQ9++feHs7Axra2s0adIEP/30U77bWbVqFSZPngwvLy84OjqiS5cuiIyMLDQ7lSJBVEJ6vV7Y2tqKVq1aFXsdPz8/Ub16dVG/fn3x888/i7///lv0799fABB79+41tbt48aIYNWqU+O2338SePXvE5s2bxeuvvy4UCoXYvXu3qV1UVJQAILy9vUWnTp3E77//LrZv3y6ioqLE1q1bhUKhEB07dhQbNmwQa9euFa1atRL+/v7i0bf+8OHDhZWVlXj33XfFtm3bxMqVK0XdunWFu7u7iI+PL/Q17dmzR1hZWYng4GCxevVqsXHjRtGtWzchSZL47bffhBBCxMbGivXr1wsAYsyYMSI8PFycOHGiwOdcsWKFACC6desmtm3bJpKTkwtsu2vXLqFWq0X79u3F6tWrxbZt28TQoUMFALFs2TJTu/PnzwtbW1tRv359sWrVKvHHH3+I7t27C19fXwFAREVFmdoCEFOmTMmzLT8/PzFkyBDT/bi4OOHj4yP8/PzEDz/8IP755x/x+eefC41GI4YOHWpql/N78vf3F88884zYuHGj2LhxowgKChLOzs7i/v37prbbtm0TVlZWolGjRmL58uVi165d4scffxQvv/xyrtfi5OQkgoKCxM8//yy2b98u3n33XaFQKMSnn35a4L562JdffikAiC1btgij0Sj8/PxEQECAMBqNudpNmTJFABCJiYm5lk+cOFEAEMePH8+zj3r16iXS09OFt7e3aN++vemx3bt3CwBi7dq1hWZ77rnnhI+PjzAYDLmWv//++0KtVos7d+4IIYQYMWKEsLW1Fd98843YvXu32Lx5s/jqq6/E999/X+jzL1u2TAAQPj4+4rXXXhNbt24VixYtEm5ubsLHx0fcu3fP1HbIkCHCzs5O6HS6fG92dna53hP5MRqNonv37kKlUomPP/5YbN++XcycOVPY2dmJpk2biszMTCGEEOHh4aJnz57CxsZGhIeHi/DwcJGQkFDg89aqVUs4ODiIWbNmiYiIiDy/uxwZGRmiUaNGws7OTsycOVNs375dfPzxx0KlUomePXvmavvoezxnXz389yHE//8ucz6Pzp8/L9q2bSs8PDxM2cPDw03tH/2bunjxonBwcBA1a9YUP//8s9iyZYt45ZVXBADx9ddf59mOv7+/ePXVV8WWLVvEqlWrhK+vr6hdu7bQ6/WF7XoqJSyWqMTi4+MFgFxfYkXx8/MT1tbWIjo62rQsIyNDuLi4iBEjRhS4nl6vFzqdTnTu3Fk899xzpuU5X8I1a9YUWq021zotWrQQPj4+Iisry7QsNTVVVK1aNVexFB4eLgCIWbNm5Vo/NjZW2NjYiPfff7/Q19S6dWvh5uYmUlNTc+Vt2LChqF69uukDPCfrjBkzCn0+IbK/XEaMGCEUCoUAICRJEvXq1RPjx4/P86Fdt25d0bRpU6HT6XIt7927t/D09DR94b700kvCxsYmV/Gn1+tF3bp1S1wsjRgxQtjb2+f6fQohxMyZMwUAcf78+VyvPSgoKNeH+5EjRwQAsWrVKtOymjVripo1a4qMjIwC90/37t1F9erV8xSRb7/9trC2thZJSUkFritE9v6tVauW8Pb2NuXJKYp27tyZq23O8vj4eKHT6cS9e/fEmjVrhJ2dnXjllVfy3Ue9evUSQgixePFiAUD8+eefQojiF0ubNm0SAMT27dtNy/R6vfDy8hIvvPCCaVnDhg1Fv379Cn2u/OQUAA//LQkhxIEDBwQA8cUXX5iWDRkyRAAo9FZUsbRt2zYBQEyfPj3X8tWrVwsAYtGiRbm2Z2dnV6zXceTIEVOxD0A4ODiI3r17i59//jlX4bRw4UIBQKxZsybX+l9//XWe/VzSYkkIIXr16iX8/Pzyzfro39TLL78sNBqNiImJydWuR48ewtbW1vQfiJztPFrUrVmzRgDIVZBR2eFhOCp3TZo0ga+vr+m+tbU16tSpg+jo6FztFi5ciGbNmsHa2hoqlQpWVlbYuXMnIiIi8jzns88+m+sk27S0NBw7dgz9+vWDWq02Lbe3t0efPn1yrbt582ZIkoT//e9/0Ov1ppuHhwcaN25s6mbPT1paGg4fPowXX3wx14moSqUSgwYNwo0bN0rUVS5JEhYuXIhr165h/vz5GDZsGHQ6Hb799ls0aNAAe/fuBQBcuXIFFy9exKuvvgoAufL37NkTcXFxpu3v3r0bnTt3hru7e66cL7300mPny7F582Z06tQJXl5eubbdo0cPADDlzNGrVy8olUrT/UaNGgGA6Xd/6dIlXL16Fa+//jqsra3z3WZmZiZ27tyJ5557Dra2tnlec2ZmZpGHvvbu3YsrV65gyJAhpjzDhg2DJEl5Dgnn8PDwgJWVFZydnTFgwAAEBwfnOWTyqGHDhqF+/fr44IMPHuscmh49esDDwwPLli0zLfv7779x69YtvPbaa6ZlLVu2xNatW/HBBx9gz549yMjIKPY2AJjeNznatGkDPz8/02HsHDY2Njh69Gi+NxsbmyK3s2vXLgDIcwi3f//+sLOzw86dOx8rd44WLVrgypUr2LZtGz788EOEhIRg586dGDx4MJ599lkIIUzbt7Ozy3MCf06ekm7/SezatQudO3fOc6HG0KFDkZ6ejvDw8FzLHz2d4dG/HSpbLJaoxFxdXWFra4uoqKjHWq9q1ap5lmk0mlwf9N988w1GjRqFVq1aYd26dTh06BCOHj2KZ555Jt8vhEfPW7h37x6EELkKgxyPLrt9+7aprZWVVa7boUOHcOfOnQJfS8528jtvwsvLC0D2+Ucl5efnh1GjRmHp0qW4fPkyVq9ejczMTLz33num7AAwceLEPNlHjx4NAKb8d+/ehYeHR55t5LesuG7fvo0///wzz7YbNGiQa9s5Hv3d55znlvM7TUxMBIBCrxq6e/cu9Ho9vv/++zzb7dmzZ77bfdTSpUsBAM899xzu37+P+/fvw8nJCe3atcO6detynUeT459//sHRo0fx999/44UXXsC///6LMWPGFLodpVKJadOm4fz580UWVg9TqVQYNGgQNmzYYMqyfPlyeHp6onv37qZ23333HSZNmoSNGzeiU6dOcHFxQb9+/XD58uVibaeg98Oj71mFQoHmzZvne3v03KD83L17FyqVCtWqVcu1XJKkfLf3OKysrNC9e3d8+eWX+PvvvxEbG4uOHTti8+bN2Lp1q2n7OUNEPMzNzQ0qleqJtl9Sd+/efazPjaL+dqhs8Wo4KjGlUonOnTtj69athV5FVBIrVqxAx44dsWDBglzLU1NT823/6Iegs7MzJEnK9+TbR8fGcXV1hSRJ2LdvX74nqRd2VZ2zszMUCgXi4uLyPHbr1i3T85eWAQMGICwsDOfOncv13KGhoXj++efzXSfn0vWqVavmOy5Qfss0Gk2+J74++gHu6uqKRo0a4csvv8x32zkf/MWV82V648aNAts4Ozubeu7eeuutfNs8fAXho5KTk01XqLVo0SLfNitXrjQVmzkaN25s2t9du3ZF9+7dsWjRIrz++usFPg8A9O3bF23btsWUKVPyXLVVmGHDhmHGjBn47bff8NJLL2HTpk0YN25crp45Ozs7TJ06FVOnTsXt27dNvUx9+vTBxYsXi9xGQe+HWrVqFTtncVStWhV6vR6JiYm5CiYhBOLj4wvdfyXZ1rhx47Bnzx6cO3cOPXv2RNWqVXH48GEIIXJ9ViQkJECv1xf6N5rTw/no30NRBXlxcpbX5wY9OfYs0RMJDQ2FEALDhw+HVqvN87hOp8Off/752M8rSVKeIuXMmTN5uqYLYmdnh+bNm2Pjxo25cj148ACbN2/O1bZ3794QQuDmzZv5/s85KCio0O20atUK69evz/U/PKPRiBUrVqB69eqoU6dOsTI/LL8P0Zz8sbGxpiIkMDAQtWvXxunTpwv8n7+DgwMAoFOnTti5c2euAtJgMGD16tV5tuPv748zZ87kWrZr1y48ePAg17LevXvj3LlzqFmzZr7bftxiqU6dOqhZsyZ+/PHHAq9SsrW1RadOnXDy5Ek0atQo3+3m13uZY+XKlcjIyMDnn3+O3bt357m5uroWeCguhyRJmDdvHpRKJT766KMiX9fXX3+N2NhY01VsxVGvXj20atUKy5Ytw8qVK5GVlYVhw4YV2N7d3R1Dhw7FK6+8gsjISKSnpxe5jV9//TXX/YMHDyI6OhodO3Ysds7iyBkqY8WKFbmWr1u3DmlpaUUOpZEfnU5XYI9QzqH6nPdf586d8eDBgzyDwP7888+58uUn56rVR/8eNm3alKftoz3khencuTN27dplKo4ezmRra8uhBioY9izREwkJCcGCBQswevRoBAcHY9SoUWjQoAF0Oh1OnjyJRYsWoWHDhnnOEypK79698fnnn2PKlCno0KEDIiMj8dlnnyEgICDXJemF+eyzz9CrVy90794d77zzDgwGA2bMmAF7e/tco4q3bdsWb775JoYNG4Zjx47hqaeegp2dHeLi4rB//34EBQVh1KhRBW4nLCwMXbt2RadOnTBx4kSo1WrMnz8f586dw6pVq/L0ehXHl19+iQMHDuCll14yDWcQFRWFuXPn4u7du5gxY4ap7Q8//IAePXqge/fuGDp0KLy9vZGUlISIiAicOHECa9euBQB89NFH2LRpE55++ml88sknsLW1xbx585CWlpZn+4MGDcLHH3+MTz75BB06dMCFCxcwd+5cODk55dnHO3bsQJs2bTB27FgEBgYiMzMT169fx19//YWFCxc+do/jvHnz0KdPH7Ru3Rrjx4+Hr68vYmJi8Pfff5u+3OfMmYN27dqhffv2GDVqFPz9/ZGamoorV67gzz//NJ0jk5+lS5fC2dkZEydOzPe8qMGDB+Obb77B6dOn0bhx4wKfp3bt2njzzTcxf/587N+/H+3atSuwbdu2bdG3b1/88ccfj7EngNdeew0jRozArVu30KZNmzwDXLZq1Qq9e/dGo0aN4OzsjIiICPzyyy8ICQmBra1tkc9/7NgxvPHGG+jfvz9iY2MxefJkeHt75+lVe1I5PXGTJk1CSkoK2rZtizNnzmDKlClo2rQpBg0a9NjPmZycDH9/f/Tv3x9dunSBj48PHjx4gD179mDOnDmoV6+eqbd18ODBmDdvHoYMGYLr168jKCgI+/fvx7Rp09CzZ0906dKlwO20aNECgYGBmDhxIvR6PZydnbFhwwbs378/T9ugoCCsX78eCxYsQHBwsOnwZX6mTJliOufvk08+gYuLC3799Vds2bIF06dPz/O3RjKT79xysiSnTp0SQ4YMEb6+vkKtVpsuCf7kk09yXfr78JVCD+vQoYPo0KGD6X5WVpaYOHGi8Pb2FtbW1qJZs2Zi48aNYsiQIbmuNinqCrMNGzaIoKAgoVarha+vr/jqq6/E2LFjhbOzc562P/74o2jVqpWws7MTNjY2ombNmmLw4MHi2LFjRb7+ffv2iaefftq0buvWrU1XQBU368MOHTok3nrrLdG4cWPh4uIilEqlqFatmnjmmWfEX3/9laf96dOnxYABA4Sbm5uwsrISHh4e4umnnxYLFy7M1e7AgQOidevWQqPRCA8PD/Hee++JRYsW5bnaJysrS7z//vvCx8dH2NjYiA4dOohTp07luVJICCESExPF2LFjRUBAgLCyshIuLi4iODhYTJ48WTx48KDI1458rrwLDw8XPXr0EE5OTkKj0YiaNWuK8ePH59mfr732mvD29hZWVlaiWrVqok2bNrmu5MpvPwEQ48aNK7DNxYsXTUM8CFHw0AFCCHH79m1hb28vOnXqZFpW0Hv8woULQqlUFutquBzJycnCxsZGABCLFy/O8/gHH3wgmjdvLpydnYVGoxE1atQQ48ePNw0tUJCcK7y2b98uBg0aJKpUqSJsbGxEz549xeXLl3O1LerqtOIMHSBE9lWvkyZNEn5+fsLKykp4enqKUaNG5RqmoDjby5GVlSVmzpwpevToIXx9fYVGoxHW1taiXr164v333xd3797N1f7u3bti5MiRwtPTU6hUKuHn5ydCQ0NNwxbkyO89funSJdGtWzfh6OgoqlWrJsaMGSO2bNmS52q4pKQk8eKLL4oqVaoISZJyXXWb3/v87Nmzok+fPsLJyUmo1WrRuHHjXMN9CFHwFZQ5f1OPtqeyIQnx3+UCRJWATqdDkyZN4O3tje3bt8sdp0JYvnw5hg0bhqioqEIHyiTLkfM7P3r0aIE9H0T0/3gYjiza66+/jq5du8LT0xPx8fFYuHAhIiIiMGfOHLmjERGRmWCxRBYtNTUVEydORGJiIqysrNCsWTP89ddfhZ6jQERE9DAehiMiIiIqBIcOICIiIioEiyUiIiKiQrBYIiIiIioET/AugtFoxK1bt+Dg4FCiwQWJiIio/AkhkJqaCi8vr2LNYVgYFktFuHXrVp5ZoYmIiMg8xMbGPvHcpSyWipAzr1ZsbCwcHR1lTkNERBWFVqvFrFmzAADvvvsu1Gq1zInoYSkpKfDx8TF9jz8JFktFyDn05ujoyGKJiIhMtFqtaX5BR0dHFksVVGmcQsMTvImIiIgKwWKJiIiIqBA8DEdERFQCCoUCjRs3Nv1MlovFEhERUQmoVCr069dP7hhUDlgKExERERWCPUtEREQlIISATqcDAFhZWXHgYgvGniUiIqIS0Ol0CAsLQ1hYmKloIsvEYomIiIioECyWiIiIiAphNsVSWFgYWrRoAQcHB7i5uaFfv36IjIwscr29e/ciODgY1tbWqFGjBhYuXFgOaYmIiMhSmE2xtHfvXrz11ls4dOgQduzYAb1ej27duiEtLa3AdaKiotCzZ0+0b98eJ0+exIcffoixY8di3bp15ZiciIiIzJnZXA23bdu2XPeXLVsGNzc3HD9+HE899VS+6yxcuBC+vr6YPXs2AKBevXo4duwYZs6ciRdeeKGsIxMREZEFMJuepUclJycDAFxcXApsEx4ejm7duuVa1r17dxw7dqzAKxeysrKQkpKS6wYABqMopeRERERkTsymZ+lhQghMmDAB7dq1Q8OGDQtsFx8fD3d391zL3N3dodfrcefOHXh6euZZJywsDFOnTs2zvOln2+HqUgUudmpUtdOgqr0arvYaeDpZo7qzLao728DHxRbOthxrg4ioMlAoFKhfv77pZ7JcZlksvf322zhz5gz2799fZNtHCxchRL7Lc4SGhmLChAmm+ykpKfDx8YFRAHceaHHngRbAgwK3Z6tWwtfFFoEeDqjj7oBAdwcEejigurMNiygiIguiUqnQv39/uWNQOTC7YmnMmDHYtGkT/v33X1SvXr3Qth4eHoiPj8+1LCEhASqVClWrVs13HY1GA41Gk2f57okdoFVY4+4DLe6mZeHOAy0SU7MQl5yBG/cyEJuUjoTULKRrDbgYn4qL8am51ne2tUJTX2c0862CZn7OaOrjDBu18jFfPREREZU3symWhBAYM2YMNmzYgD179iAgIKDIdUJCQvDnn3/mWrZ9+3Y0b94cVlZWj7X9ag7WcHR0LLRNps6Am/czEJWYhsjbqbh8OxWRtx/gasID3EvXYdfFBOy6mAAAUKsUaBXggg51qqFjoBtqudk/Vh4iIiIqH5LIOS5VwY0ePRorV67EH3/8gcDAQNNyJycn2NjYAMg+hHbz5k38/PPPALKHDmjYsCFGjBiB4cOHIzw8HCNHjsSqVauKfTVcSkoKnJyckJycXGSxVBCt3ogLcSk4EX0PJ2Lu4Xj0PcQlZ+ZqU8fdHn2beOPZxl7wcbEt0XaIiKj8aLVahIWFAcj+/lGr1TInooeVxvd3DrMplgo632fZsmUYOnQoAGDo0KG4fv069uzZY3p87969GD9+PM6fPw8vLy9MmjQJI0eOLPZ2S3Nn5xBC4GpiGvZEJmDvpUQcvpYErcFoeryFvzMGhfjjmQYeUKt40iARUUXEYqliq5TFklzKolh6VHKGDn+fi8em07dw8Ood5IxS4Oagwaut/DC0jT+cbB/vsCEREZUtFksVW2l+f5vNOUuWzMnGCgNa+GBACx8kpGRi5ZEY/Ho4BgmpWfj2n0tYsu8ahrULwOvtAuBkw6KJiIioPPEYTwXj5miNcV3q4MCkpzHn5SYIdHdAapYe3+28jKem78Yv4dc5QCYREVE5YrFUQalVCvRt4o2t77THvIHNUNvNHskZOnz8x3n0/n4/jl1PkjsiERFRpcBiqYJTKCT0auSJre+0x2d9G8DJxgoRcSno/0M4vth8AZk6g9wRiYiILBqLJTOhUiowOMQfuyd2RP/g6hACWLI/Cr2/349zN5PljkdEVOkoFArUrl0btWvX5nQnFo5XwxWhPK6GK4mdEbfxwfqzSEzNglqlwBd9G2JACx+5YxEREVUIpfn9zVLYTHWu547t455C57pu0OqNeH/dGYSuP4ssPQ/LERERlSYWS2bM2U6NxYObY0LXOpAkYNWRGAz58QhSMnVyRyMiIrIYLJbMnEIhYWzn2vhxaAvYa1Q4dC0JAxaG43ZKZtErExFRiWm1WkybNg3Tpk2DVquVOw6VIRZLFqJToBtWj2iNag4aXIxPxfPzD+L6nTS5YxERWTSdTgedjr35lo7FkgVp4OWE9aPaIMDVDjfvZ+CVxYcQczdd7lhERERmjcWShfFxscWaESGo5WaPuORMvLL4EGKTWDARERGVFIslC1TNQYOVw1uhRrXsHqZXlxzGnQdZcsciIiIySyyWLJSbgzVWDW8NXxdbxCSl4/XlR5Gu1csdi4iIyOywWLJg7o7WWD6sBZxtrXD6RjLGrDwJvcEodywiIiKzwmLJwtWoZo8lQ1pAo1Jg58UEfLb5gtyRiIgsgiRJ8PPzg5+fHyRJkjsOlSFOd1KEijrdyePadi4eI1ccBwDMeLER+jfn1ChERGS5ON0JPbZnGnpgXJfaAIDJG8/h7A1OvktERFQcLJYqkbFP1zbNJTdyxXEkpXHEWSIioqKwWKpEFAoJ37zUxDRo5btrToFHYYmISkar1WLGjBmYMWMGpzuxcCyWKhknGyss+F8zqFUK7I5MxE8Hr8sdiYjIbKWnpyM9nQP/WjoWS5VQXQ9HTO5ZDwAwbetFXIxPkTkRERFRxcViqZIaHOKHp/87f2nsqpPI1BnkjkRERFQhsViqpCRJwvQXG8HVXoNLtx9g1vZIuSMRERFVSCyWKjFXew2mvxgEAFi6PwonYu7JnIiIiKjiYbFUyT1d1x3PN/WGUQDv/34GWXoejiMiInoYiyXCJ33qw9VegysJD/D9zityxyEiMguSJMHLywteXl6c7sTCcbqTIljKdCdF2XYuDiNXnIBSIeGPt9qiobeT3JGIiIhKjNOdUKl7pqEnegV5wmAU+HDDWRiMrKGJiIgAFkv0kCnP1oeDRoUzN5Kx6kiM3HGIiIgqBBZLZOLmYI0J3eoAAGb8HYm7D7JkTkREVHHpdDrMnj0bs2fPhk6nkzsOlSEWS5TLoNZ+qO/piOQMHb7aelHuOEREFZYQAsnJyUhOTuY8mxaOxRLlolIq8Hm/hgCAtcdv4Nj1JJkTERERyYvFEuUR7OeMAc2rAwA+2ngOeoNR5kRERETyYbFE+fqgRz042VjhYnwqfjsaK3ccIiIi2bBYony52KkxvkttAMC3Oy4hJZMnLxIRUeXEYokK9GprP9SsZoe7aVrM28WRvYmIqHJisUQFslIqMLlXPQDAsgPXEXM3XeZEREQVhyRJqFatGqpVq8bpTiwcpzspQmWZ7qQgQggM/vEI9l2+gx4NPbDgf8FyRyIiIioSpzuhciNJEj7qVR8KCdh6Lh6Hr92VOxIREVG5YrFERQr0cMDLLX0BAF9siYCR88YREVElwmKJimVC1zpw0Khw9mYy1p+8KXccIiLZ6XQ6zJ8/H/Pnz+d0JxaOxRIVi6u9Bm89XQsAMPPvSGRoDTInIiKSlxACiYmJSExM5HQnFo7FEhXb0Db+8K5ig/iUTCzdf03uOEREROWCxRIVm7WVEu8/EwgAWLDnKhJSM2VOREREVPbMqlj6999/0adPH3h5eUGSJGzcuLHQ9nv27IEkSXluFy9eLJ/AFqhPIy80ru6ENK0Bs/+5LHccIiKiMmdWxVJaWhoaN26MuXPnPtZ6kZGRiIuLM91q165dRgktn0IhYXKv+gCA347E4NLtVJkTERERlS2V3AEeR48ePdCjR4/HXs/NzQ1VqlQp/UCVVMsAF3Rv4I6/z99G2F8RWDaspdyRiIiIyoxZ9SyVVNOmTeHp6YnOnTtj9+7dcsexCB/0qAeVQsLuyETsv3xH7jhEROVOkiQ4OTnBycmJ051YOIsuljw9PbFo0SKsW7cO69evR2BgIDp37ox///23wHWysrKQkpKS60Z5Bbja4X+t/QAAX/4VAQMHqiSiSsbKygrjxo3DuHHjYGVlJXccKkNmdRjucQUGBiIwMNB0PyQkBLGxsZg5cyaeeuqpfNcJCwvD1KlTyyuiWXunc22sO3EDEXEpWH/iBvo395E7EhERUamz6J6l/LRu3RqXLxd8FVdoaCiSk5NNt9jY2HJMZ16c7dQYkzNQ5fZIpGv1MiciIiIqfZWuWDp58iQ8PT0LfFyj0cDR0THXjQo2pI0/qjvb4HZKFpbsi5I7DhFRudHpdFi8eDEWL17M6U4snFkdhnvw4AGuXLliuh8VFYVTp07BxcUFvr6+CA0Nxc2bN/Hzzz8DAGbPng1/f380aNAAWq0WK1aswLp167Bu3Tq5XoLF0aiUmPRMXYxZdRIL917Fyy184OZoLXcsIqIyJ4TArVu3TD+T5TKrnqVjx46hadOmaNq0KQBgwoQJaNq0KT755BMAQFxcHGJiYkzttVotJk6ciEaNGqF9+/bYv38/tmzZgueff16W/JaqdyNPNPGpgnStAd/+c0nuOERERKVKEiyHC5WSkgInJyckJyfzkFwhjkcn4YUF4VBIwNZ3nkKgh4PckYiIypRWq0VYWBiA7PNd1Wq1zInoYaX5/W1WPUtUcQX7uaBnkAeMApj2V4TccYiIiEoNiyUqNe93rwsrpYS9lxLx76VEueMQERGVChZLVGr8Xe0wqLU/gOzeJQ5USUREloDFEpWqsZ1rwdFahYvxqVh3/IbccYiIypStrS1sbW3ljkFljCd4F4EneD++Jfuu4YstEXBz0GD3xI6w05jVCBVERGQBeII3VWiDQvzg62KLhNQsLN53Te44RERET4TFEpW6nIEqAeCHvddwOyVT5kREREQlx2KJykTPIA80862CDJ0B32znQJVEZHl0Oh2WL1+O5cuXc7oTC8diicqEJEmY3Ks+AGDN8VhExKXInIiIqHQJIRAdHY3o6GhOd2LhWCxRmQn2c0avRp4QHKiSiIjMGIslKlOTuteFWqnAvst3sDsyQe44REREj43FEpUp36q2GNLGDwAwddN5ZOoMMiciIiJ6PCyWqMyN7Vwb7o4aXL+bjnm7r8gdh4iI6LGwWKIy52BthanPNgAALNx7FZdvp8qciIiIqPhYLFG56N7AA13quUFnEPhww1kYOW8cEVkAKysrWFlZyR2DyhinOykCpzspPTfvZ6DrN3uRrjXg6xeC8FILX7kjERGRheJ0J2SWvKvYYELXOgCAaX9dRAJH9iYiIjPAYonK1dA2/mjo7YjkDB0mrTvDgdyIiKjCY7FE5UqlVGBW/yZQKxXYHZmINcdi5Y5ERFQier0eK1euxMqVK6HX6+WOQ2WIxRKVu0APB7zbLftw3Gd/XkBsUrrMiYiIHp/RaMTly5dx+fJlGI1GueNQGWKxRLJ4o30NtPB3RprWgIlrT/PqOCIiqrBYLJEslAoJM/s3hq1aicNRSVj471W5IxEREeWLxRLJxq+qHT7tkz1Y5aztl3AkKknmRERERHmxWCJZ9W9eHc839YbBKDBm1QncfZAldyQiIqJcWCyRrCRJwhfPNUQtN3vcTsnCuNWnYOD5S0REVIGwWCLZ2apVmDewGaytFNh3+Q6+3nZR7khEREQmnO6kCJzupPz8efoWxqw6CQCY/mIjDGjuI3MiIiIyV5zuhCxSn8ZeGNu5NgBg8oazPOGbiIgqBBZLVKGM61wbvYI8oTMIvPnLMVy6nSp3JCIiquRYLFGFovhv/KUmPlVwP12HQUsPc4RvIqqQ9Ho91q5di7Vr13K6EwvHYokqHBu1EsuHtUAd9+wr5F5dchgJKZlyxyIiysVoNOLChQu4cOECpzuxcCyWqEKqYqvGL6+3gq+LLWKS0vHy4kOIS86QOxYREVVCLJaownJ3tMaK11vB08ka1xLT0H9hOGLu8pAcERGVLxZLVKH5VrXF2pEh8K9qixv3MvDiwoOIjOdJ30REVH5YLFGFV93ZFmtGhiDQ3QEJqVl4YcFB7Lp4W+5YRERUSbBYIrPg5mCN1SNao1WACx5k6fH6T8fww96r4JiqRERU1lgskdnIOen7lZY+EAII23oRo1acwP10rdzRiIjIgnG6kyJwupOKRwiBnw5exxdbIqA3Cng4WuPbl5ogpGZVuaMRUSUihIBOpwMAWFlZQZIkmRPRwzjdCVVqkiRhaNsArB/dBgGudohPycTAJYfw8cZzSMnUyR2PiCoJSZKgVquhVqtZKFk4FktkthpVr4LNY9rhpebZh+V+ORSNzrP2YtPpWzyXiYiISg0PwxWBh+HMw8Erd/DRxnO4dicNANCouhMmPVMXbWu5ypyMiCyVXq/H5s2bAQC9e/eGSqWSORE9jIfhiB7RppYr/nqnPSZ0rQM7tRJnbiTj1SWH8fKicOyOTGBPExGVOqPRiNOnT+P06dOc7sTCsQwmi2FtpcTYzrUxsJUv5u66gl8PR+PQtSQcupaE2m72eK1dAPo09oK9hm97IiIqPvYskcVxtdfg02cbYO97nTC8fQDsNSpcTniA0PVn0eKLfzBh9SkcvHIHegP/J0hEREXjf7HJYnlVscHkXvUxpnNtrD4Si1VHY3AtMQ3rT97E+pM3UcXWCk/XdUPXeu5oV9sVDtZWckcmIqIKiMUSWTxHaysMf6oG3mgfgJOx97H22A1sPReH++k6rD9xE+tP3IRCAoK8ndCqRlW0CnBBU19nuNip5Y5OREQVgFkVS//++y9mzJiB48ePIy4uDhs2bEC/fv0KXWfv3r2YMGECzp8/Dy8vL7z//vsYOXJk+QSmCkWSJDTzdUYzX2d83rcBjkXfwz8XbuOfiNu4fjcdp28k4/SNZCz69xoAwMvJGvW9nNDQ2xF1PRzg72oHPxc72KiVMr8SIiIqT2ZVLKWlpaFx48YYNmwYXnjhhSLbR0VFoWfPnhg+fDhWrFiBAwcOYPTo0ahWrVqx1ifLpVIq0LpGVbSuURUf9a6PW/czcDjqLg5fS8KRqCRcu5OGW8mZuJWciX8ick/a6+lkDf+qdqjubAMPJ2u4O2bfPByt4e6kgbOtGlZKng5IRGQpzHacJUmSiuxZmjRpEjZt2oSIiAjTspEjR+L06dMIDw8v1nY4zlLllJqpQ0RcKs7dTMb5Wym4kvgAUYkPkJKpL9b69hoVnGysUMX2v5uNGo42VrBVK2GnVsJGrYKtWgkbtfK/ZSrYqJXQqBSwUiqgVimgUkimn62UClgppf/+VUCp4GjBRHITQiA9PR0AYGtry1G8K5jS/P42q56lxxUeHo5u3brlWta9e3csXboUOp0OVlZ5T+jNyspCVlaW6X5KSkqZ56SKx8HaCi0DXNAywMW0TAiBe+k6RN1Jw/U7abh1PwPxKZm4nZKF2ymZiE/JxJ0HWRACeJClx4MsPW7ezyiTfAoJpqJJIUmQJJh+Vkj4718JSkX2Yw//rJTyXwcPrSsBpjbSQ8+XfT+73f+3ydkGIEGCQpHdJudxxX/rSI88t0IBAA9t879/kWu9vNvIeQ7poUwFbePhvHnz//9ySAVsE4BKIcHaSgmNlQI2VkpYm24KWKuUpiKXX5SVjyRJsLOzkzsGlQOLLpbi4+Ph7u6ea5m7uzv0ej3u3LkDT0/PPOuEhYVh6tSp5RWRzIgkSXCxU8PFTo1gP+d82xiMAikZOtzP0OF+uhb3M3RITs/+OTlDj3SdHulZBqRrDcjQ6ZGu/e9nrQFpWj20eiP0BgGdwQitwQidwQidQcBgzN0BbBRAlp5DH1Qk1lYKOFpbwdHGCo7W2T2L2T9bmXoZqzlo4OZgDTdHDdwdrTnmF5GZsPi/1Ef/t5dz1LGg/wWGhoZiwoQJpvspKSnw8fEpu4BkUZQKCc52ajjbqQGU3v84DUbxX+GUXTzpDEZo9UYIARiFgEEICCFgFNltjUJAPPSzUWS/97Pv//fzf8uNRgEBAaMREMB/64r/nvu/+8B/z597Of771yiQ/RzZDf9/vYf+zXncdN+U66Ft4qHn+6+NQHbGnG2IR547u47Mzm986Dke3sbDr+Hh9f6/Tc56j7bJ3meZOiMydAZk6gzI1BmRpTMgQ2eA/qEiNlNnRKYuCwmpWXl+fwWxVSvh4WgN36q28K9qB7+H/vWrasfDrRWcXq/H33//DSD7qAWnO7FcFv2b9fDwQHx8fK5lCQkJUKlUqFq1ar7raDQaaDSa8ohHVGxKhQSlIvvwD1UceoMRmXojMrTZhVRyhg4pmTqkZOiR8t/PyRk6pGTokJSuQ0JKJhJTswuqB1nZPYvX7qT9N6dhYq7ntrFSoq6nAxp4OaKBlxOCvJ1Qz9ORBVQFYjQacezYMQBA165dZU5DZcmii6WQkBD8+eefuZZt374dzZs3z/d8JSKix6FSKmCvVJgOpz1OH3Ralh4JqVmIu5+B6KR0XL+Thut30xB9Nx3X76YhQ2fAyZj7OBlz37SOg7UKLfxd0LqGC9rUdEUDL0eeK0VUDsyqWHrw4AGuXLliuh8VFYVTp07BxcUFvr6+CA0Nxc2bN/Hzzz8DyL7ybe7cuZgwYQKGDx+O8PBwLF26FKtWrZLrJRARAQDsNCoEaFQIcLVDm0ceMxgFou6k4fytZFy4lYLzt1JwOvY+UjP12HUxAbsuJgDIHsaiSz13dK3vjpCaVTlkBVEZMati6dixY+jUqZPpfs65RUOGDMHy5csRFxeHmJgY0+MBAQH466+/MH78eMybNw9eXl747rvvOMYSEVVoSoWEWm72qOVmj75NvAFkF1AXbqXgcNRdHLp2Fwev3kVcciZ+ORSNXw5Fw9Veg+eaeuHFYB8EejjI/AqILIvZjrNUXjjOEhFVRJk6Aw5evYMdF25j+/nbuJumNT0W7OeM4e0D0LW+B89xKkNarRZhYWEAsi8OUqs5RVJFwnGWiIgqOWsrJZ6u646n67rjs75G7IlMxO/HY7EzIgHHo+/hePQ9+FW1xfD2NfBSCx8eoiN6AiyWiIjMnJVSga71s89dSkjNxM8Hsw/NRd9Nx0cbz2Hxvmt4t1sgegd5QsGeJqLHxsNwReBhOCIyR+laPX47Eov5e67gzoPsQ3SNfargy34N0dDbSeZ0lkEIgeTkZACAk5MTr0ysYErz+5vFUhFYLBGROUvL0uPH/VH44d9reJClh0IChrUNwISudWDHEcTJgpXm9zcPYhMRWTA7jQpjOtfGrnc7oFcjTxgFsHR/FHp9tw+nY+/LHY/ILLBYIiKqBNwcrTFvYDMsG9YCXk7WuH43HS8sOIiFe6/CaOQBhpIwGAzYvn07tm/fDoPBIHccKkMsloiIKpFOgW7Y+s5T6BnkAb1R4KutF/HmL8fwIEsvdzSzYzAYEB4ejvDwcBZLFo7FEhFRJeNka4V5A5vh6xeCoFYp8E9EAl5ccBCxSelyRyOqkFgsERFVQpIk4aUWvlgzIgTVHDS4GJ+KvvMO4BTPYyLKg8USEVEl1sSnCja93RZB3k5IStPi1cWHcPDqHbljEVUoLJaIiCo5Tycb/PZma7StVRVpWgOGLjuKnRG35Y5FVGGwWCIiIthpVFg6pAW61neHVm/EqBUn8O+lRLljEVUILJaIiAhA9nxzC15thp5BHtAajHjzl2M4ej1J7lhEsuMI3kXgCN5EVNlo9UYM//kY9l5KhINGhZXDWyOoOqdIeZQQAomJ2b1v1apV43QnFQxH8CYiojKjVimw8H/BaBnggtQsPV776Shu3c+QO1aFI0kS3Nzc4ObmxkLJwrFYIiKiPGzUSiwd0hyB7g5ITM3C6z8dQxoHrqRKisUSERHly8HaCkuGNIervRoRcSl457dTMHBqFBODwYA9e/Zgz549HMHbwrFYIiKiAvm42OKHQc3/G+n7Nr7dcUnuSBWGwWDA3r17sXfvXhZLFo7FEhERFSrYzxnTX2gEAJi7+wp2RybInIiofLFYIiKiIvVr6o3/tfYFAIxffYonfFOlwmKJiIiK5ePe9RHk7YT76Tq8tfIEtHqj3JGIygWLJSIiKhaNSol5A5vBwVqFkzH3MWcnz1+iyoHFEhERFZtvVVvT+UsL9lzF8WiO8E2Wj8USERE9lh5Bnni+qTeMApiw5jTHXyKLx2KJiIge26d9G8DLyRrRd9Px5V8RcseRhUqlwhtvvIE33ngDKpVK7jhUhlgsERHRY3O0tsLM/o0BACsPx2Df5USZE5U/hUIBb29veHt7Q6Hg16kl42+XiIhKpE0tVwxt4w8AmLzhHDK0HJiRLNMTFUtarRaRkZHQ63m8moioMprYPRBeTtaISUrH7H8q19VxBoMBBw4cwIEDBziCt4UrUbGUnp6O119/Hba2tmjQoAFiYmIAAGPHjsVXX31VqgGJiKjisteo8Hm/hgCAJfujcO5mssyJyo/BYMA///yDf/75h8WShStRsRQaGorTp09jz549sLa2Ni3v0qULVq9eXWrhiIio4utczx29GnnCYBQIXX8WegMHqyTLUqJiaePGjZg7dy7atWsHSZJMy+vXr4+rV6+WWjgiIjIPU/rUh6O1CmdvJuOXQ9FyxyEqVSUqlhITE+Hm5pZneVpaWq7iiYiIKgc3B2t80KMeAOCbHZdw50GWzImISk+JiqUWLVpgy5Ytpvs5BdLixYsREhJSOsmIiMisvNTCBw29HZGaqcf0bRfljkNUako0ilZYWBieeeYZXLhwAXq9HnPmzMH58+cRHh6OvXv3lnZGIiIyA0qFhKnPNsQLCw5izbEbGNjKD018qsgdi+iJlahnqU2bNjhw4ADS09NRs2ZNbN++He7u7ggPD0dwcHBpZyQiIjMR7OeMF5pVBwBM+eMcjEYhcyKiJycJIfhOLkRKSgqcnJyQnJwMR0dHueMQEVV4CamZeHrmXjzI0mP6C40woIWP3JHKhNFoNA2d4+vry1G8K5jS/P4u0W82JSUl31tqaiq0Wu0TBSIiIvPm5mCNcV1qAwC+3nYRqZk6mROVDYVCAX9/f/j7+7NQsnAl+u1WqVIFzs7OeW5VqlSBjY0N/Pz8MGXKFBiNHGuDiKgyGtLGHzWq2eFumhYL93JIGTJvJSqWli9fDi8vL3z44YfYuHEjNmzYgA8//BDe3t5YsGAB3nzzTXz33XcczZuIqJKyUioQ+t9QAkv2ReHW/QyZE5U+g8GAI0eO4MiRIxzB28KV6Gq4n376CbNmzcKAAQNMy5599lkEBQXhhx9+wM6dO+Hr64svv/wSH374YamFJSIi89GlnhtaBbjgcFQSZv4diW9eaiJ3pFJlMBiwdetWAECTJk2gVCplTkRlpUQ9S+Hh4WjatGme5U2bNkV4eDgAoF27dqYT34iIqPKRJAmTe2X3Lq0/eRNnb1SeeePIspSoWKpevTqWLl2aZ/nSpUvh45N91cPdu3fh7Oz8ZOmIiMisNapeBc819QYAfPnXBfACbDJHJToMN3PmTPTv3x9bt25FixYtIEkSjh49iosXL+L3338HABw9ehQvvfRSqYYlIiLzM7F7ILacjcOha0nYGZGALvXd5Y5E9FhK1LP07LPP4tKlS+jZsyeSkpJw584d9OjRAxcvXkTv3r0BAKNGjcI333xTqmGJiMj8eFexwevtAgAA07ZGQGfgldJkXh67Z0mn06Fbt2744YcfEBYWVhaZiIjIwozuWBNrjsbiWmIaVh2JweAQf7kjERXbY/csWVlZ4dy5c6bJc4mIiIriYG1lGqhyzj+XLXagSrJMJToMN3jw4HxP8C4P8+fPR0BAAKytrREcHIx9+/YV2HbPnj2QJCnP7eJFzoZNRFTeXm7pixqu2QNVLvr3mtxxnphKpcIrr7yCV155BSpViU4BJjNRot+uVqvFkiVLsGPHDjRv3hx2dna5Hi+rc5VWr16NcePGYf78+Wjbti1++OEH9OjRAxcuXICvr2+B60VGRuaaF6ZatWplko+IiApmpVTg/WfqYuSK41i87xpebeUHDydruWOVmEKhQJ06deSOQeWgRBPpdurUqeAnlCTs2rXriUIVpFWrVmjWrBkWLFhgWlavXj3069cv3/On9uzZg06dOuHevXuoUqVKibbJiXSJiEqPEAL9F4bjWPQ9vNTcB1+/2EjuSGShSvP7u0Q9S7t3736ijZaEVqvF8ePH8cEHH+Ra3q1bNxw8eLDQdZs2bYrMzEzUr18fH330UaHFXlZWFrKyskz3U1JSniw4ERGZSJKE0J718MKCg1h7PBavtQtAoIeD3LFKxGAw4OzZswCAoKAgjuBtwcxmmuQ7d+7AYDDA3T33+Bzu7u6Ij4/Pdx1PT08sWrQI69atw/r16xEYGIjOnTvj33//LXA7YWFhcHJyMt1yBtkkIqLSEeznjB4NPWAUwFdbI+SOU2IGgwF//PEH/vjjD84NZ+FKfEba0aNHsXbtWsTExECr1eZ6bP369U8crCCPXoUnhCjwyrzAwEAEBgaa7oeEhCA2NhYzZ87EU089le86oaGhmDBhgul+SkoKCyYiolL2/jN1sePCbeyOTMTBK3fQppar3JGIClSinqXffvsNbdu2xYULF7BhwwbodDpcuHABu3btgpOTU2lnBAC4urpCqVTm6UVKSEjI09tUmNatW+Py5csFPq7RaODo6JjrRkREpSvA1Q6vtsq+MGfa1ggYjZwGhSquEhVL06ZNw7fffovNmzdDrVZjzpw5iIiIwIABAwq9Ku1JqNVqBAcHY8eOHbmW79ixA23atCn285w8eRKenp6lHY+IiB7T2M61Ya9R4dzNFPx55pbccYgKVKJi6erVq+jVqxeA7J6YtLQ0SJKE8ePHY9GiRaUa8GETJkzAkiVL8OOPPyIiIgLjx49HTEwMRo4cCSD7ENrgwYNN7WfPno2NGzfi8uXLOH/+PEJDQ7Fu3Tq8/fbbZZaRiIiKp6q9BqM61gQATN8WiUwdz/uhiqlE5yy5uLggNTUVAODt7Y1z584hKCgI9+/fR3p6eqkGfNhLL72Eu3fv4rPPPkNcXBwaNmyIv/76C35+fgCAuLg4xMTEmNprtVpMnDgRN2/ehI2NDRo0aIAtW7agZ8+eZZaRiIiK77W2AfglPBo372fgl/BoDH+qhtyRiPIo0ThLAwcORPPmzTFhwgR8+eWXmDNnDvr27YsdO3agWbNmZXqCd3njOEtERGVrzbFYvP/7GThaq/Dv+51QxVYtd6Ri0Wq1pjH+QkNDoVabR+7KojS/v0tULCUlJSEzMxNeXl4wGo2YOXMm9u/fj1q1auHjjz+Gs7PzE4WqSFgsERGVLYNRoOecfYi8nYrh7QMwuVd9uSMVi9FoRERE9tAH9erVg0JhNqPxVAqyF0uVCYslIqKytzsyAcOWHYVaqcDOdzvAx8VW7khk5mQfwRvIrqivXLmChIQEGI3GXI8VNIYRERFRfjrWqYa2tariwJW7mLk9EnNebip3JCKTEhVLhw4dwsCBAxEdHY1HO6YkSeJIpkRE9FgkSUJoj3ro/f1+/HHqFt5oVwNB1ctm3L7SwsNwlUeJfrMjR45E8+bNce7cOSQlJeHevXumW1JSUmlnJCKiSqChtxOea+oNAJj2V0Se/4xXNHq9Hr///jt+//136PV6ueNQGSpRz9Lly5fx+++/o1atWqWdh4iIKrF3u9XBlrNxCL92F3siE9GprpvckYhK1rPUqlUrXLlypbSzEBFRJVfd2RbD2vgDAMK2RsDAaVCoAih2z9KZM2dMP48ZMwbvvvsu4uPjERQUBCsrq1xtGzVqVHoJiYioUhndsRZ+OxqLS7cf4PfjsXipRdlMo0VUXMUulpo0aQJJknIdQ37ttddMP+c8xhO8iYjoSTjZWmHM07XwxZYIzNx+CT2DPOFgbVX0ikRlpNjFUlRUVFnmICIiMhkU4odfD8cg6k4a5vxzGR/1No+BKskyFfucJT8/P9Nt5cqV2LlzZ65lfn5+2LlzJ3777beyzEtERJWARqXElD7ZBdKyg9dx6XaqzImoMivRCd4//PAD6tatm2d5gwYNsHDhwicORURE1DHQDV3ru8NgFJjyx/kKN5SAUqlE37590bdvXyiVSrnjUBkqUbEUHx8PT0/PPMurVauGuLi4Jw5FREQEAJ/0rg+NSoHwa3ex+UzF+n5RKpVo0qQJmjRpwmLJwpWoWPLx8cGBAwfyLD9w4AC8vLyeOBQREREA+LjYYnTH7DH9vtwSgbQsDv5I5a9ExdIbb7yBcePGYdmyZYiOjkZ0dDR+/PFHjB8/HsOHDy/tjEREVImN6FADvi62iE/JxDc7Lskdx8RoNOLSpUu4dOlSnjlSybKUaATv999/H0lJSRg9ejS0Wi0AwNraGpMmTUJoaGipBiQiosrN2kqJz/o2wNBlR7HsQBT6NPZCE58qcseCXq/HqlWrAAChoaFQq9UyJ6KyUqKeJUmS8PXXXyMxMRGHDh3C6dOnkZSUhE8++aS08xEREaFjoBuea+oNowAm/X4GWj17cqj8PNEUyfb29mjRogUaNmwIjUZTWpmIiIjy+Lh3fbjYqRF5OxUL916VOw5VIk9ULBEREZUXFzu1aeylubuu4DLHXqJywmKJiIjMxrONvdApsBq0BiPeXXsaOgMPx1HZY7FERERmQ5IkTHs+CI7WKpy5kYzvdl6WOxJVAiyWiIjIrHg62WDa80EAgHm7r+DY9SSZE5GlY7FERERmp3cjLzz/39Vx49ecQmqmrtwzKJVK9OjRAz169OAI3haOxRIREZmlqX0boLqzDWKTMmSZO06pVKJly5Zo2bIliyULx2KJiIjMkoO1Fb59qQkUErD+5E2sOhIrdySyUCyWiIjIbLXwd8H7z9QFAHy66TxOx94vt20bjUZcv34d169f53QnFo7FEhERmbURT9VAt/ru0BqMGP3rCdxL05bLdvV6PX766Sf89NNP0Os5wa8lY7FERERmTZIkzBzQGP5VbXHzfgZG/3qC06FQqWKxREREZs/R2goLBwXDTq1E+LW7+Gjj2XI/4ZssF4slIiKyCHU9HDF3YDMoJGDNsRtYwPnjqJSwWCIiIovRqa4bPn22AQBg+rZI/HHqpsyJyBKwWCIiIosyOMQfw9r6AwAmrDmNHRduyxuIzB6LJSIisjgf96qP55t6w2AUeOvXE9h/+Y7ckciMsVgiIiKLo1BImP5iIzzTwANagxHDfz6Gg1dKt2BSKpXo0qULunTpwhG8LRyLJSIiskgqpQJzXmmCDnWqIUNnwNDlR/FPKR6SUyqVaNu2Ldq2bfvExVLHjh0xbty40gmWj8zMTAwdOhRBQUFQqVTo169fnjZxcXEYOHAgAgMDoVAoip0nJiYGffr0gZ2dHVxdXTF27FhotbnHulqzZg2aNGkCW1tb+Pn5YcaMGaXwqsoPiyUiIrJYGpUSPwwKRtf67tDqjRi54jg2nb4ld6xyZzAYYGNjg7Fjx6JLly75tsnKykK1atUwefJkNG7cuNjP26tXL6SlpWH//v347bffsG7dOrz77rumNlu3bsWrr76KkSNH4ty5c5g/fz6++eYbzJ07t1ReW3mQBAeiKFRKSgqcnJyQnJwMR0dHueMQEVEJ6AxGvLf2NDaeugVJAiY9UxcjnqoBSZJK/JxGoxFxcXEAAE9PTygU+fc/dOzYEQ0bNgQArFixAkqlEqNGjcLnn39u2n7Hjh3RpEkTzJ49u8R5imvo0KG4f/8+Nm7cWGCb4ubZunUrevfujdjYWHh5eQEAfvvtNwwdOhQJCQlwdHTEwIEDodPpsHbtWtN6s2fPxqxZsxATE/NEv4PClOb3N3uWiIjI4lkpFfhmQBMMbeMPIYCvtl7E+7+feaKRvvV6PZYsWYIlS5YUOd3JTz/9BJVKhcOHD+O7777Dt99+iyVLlhR7W/v27YO9vX2ht2nTppX4tZRUeHg4GjZsaCqUAKB79+7IysrC8ePHAWT3WFlbW+daz8bGBjdu3EB0dHS55i0pldwBiIiIyoNCIeHTZxsgwNUOU/88j7XHbyA6KR3zX20GV3tNmW7bx8cH3377LSRJQmBgIM6ePYtvv/0Ww4cPL9b6zZs3x6lTpwpt4+LiUgpJH098fDzc3d1zLXN2doZarUZ8fDyA7OJp/PjxGDp0KDp16oQrV66Yeqzi4uLg7+9fzqkfH4slIiKqVIa08YdfVVuMWXkSR6KS0HPOPsx5uSlCalYts222bt061+GmkJAQzJo1CwaDoVgnh9vY2KBWrVpllu9J5HcYTQhhWj58+HBcvXoVvXv3hk6ng6OjI9555x18+umnZnMVIQ/DERFRpdMx0A0b3mqDOu72SEjNwqtLDmHOP5dhMFbM03gr6mE4Dw8PUw9Sjnv37kGn05l6nCRJwtdff40HDx4gOjoa8fHxaNmyJQCYRa8SwJ4lIiKqpGq5OeCPt9phyqZzWHPsBr795xL2XU7E9BcboUY1+1Ld1qFDh/Lcr127drF7VirqYbiQkBB8+eWXiIuLg6enJwBg+/bt0Gg0CA4OztVWqVTC29sbALBq1SqEhITAzc2t3DOXBIslIiKqtGzUSkx/sTFaBVTFlE3ncSz6HnrM2Yf3n6mLYW38oVCUzpVasbGxmDBhAkaMGIETJ07g+++/x6xZs4qfsxQOw124cAFarRZJSUlITU01FV9NmjQxtclZ9uDBAyQmJuLUqVNQq9WoX78+AGDDhg0IDQ3FxYsXAQDdunVD/fr1MWjQIMyYMQNJSUmYOHEihg8fbroC7c6dO/j999/RsWNHZGZmYtmyZVi7di327t37RK+nPLFYIiKiSu+F4OpoVcMFH6w7i/1X7uDzzRew9WwcvniuIep6PPmwMYMHD0ZGRgZatmwJpVKJMWPG4M033yyF5MXXs2fPXFefNW3aFED2+UWPLgOA48ePY+XKlfDz88P169cBAMnJyYiMjDS1USqV2LJlC0aPHo22bdvCxsYGAwcOxMyZM3Nt+6effsLEiRMhhEBISAj27NljOhRnDjjOUhE4zhIRUeUhhMDKIzGYtiUCaVoDlAoJQ0L8Mb5rbThYW+VqazAYsG/fPgBA+/btCzykVp5jKNH/q9TjLM2fPx8BAQGwtrZGcHCw6Y1akL179yI4OBjW1taoUaMGFi5cWE5JiYjI3EiShFdb+WHHhA7o0dADBqPAjwei8PSsvdh48mauXhilUomOHTuiY8eOZnNVF5WMWRVLq1evxrhx4zB58mScPHkS7du3R48ePRATE5Nv+6ioKPTs2RPt27fHyZMn8eGHH2Ls2LFYt25dOScnIiJz4lXFBgv+F4yfXmuJAFc7JKZmYdzqU3hu/kEcu54kdzwqZ2Z1GK5Vq1Zo1qwZFixYYFpWr1499OvXD2FhYXnaT5o0CZs2bUJERIRp2ciRI3H69GmEh4cXa5s8DEdEVLll6Q1Y/O81zN9zFelaAwCgR0MPTHomELbGdABAtWrVymzaDiqZ0vz+NpsTvLVaLY4fP44PPvgg1/Ju3brh4MGD+a4THh6Obt265VrWvXt3LF26FDqdDlZWVvmulx+j0QijseTD4hMRkXmyUkgY3bEmXmzmjW//uYy1x29g67l47I64hVfUJwBk/+dcrVbLnJQeVprf2WZTLN25cwcGgyHPsOru7u55BsTKkd8w7O7u7tDr9bhz545pTIiHZWVlISsry3Q/JSUFAHDx4kXY25fuuBtERGReBtVVor27J5advIfTt9JMy6dvPIo+9arASsnepYriwYMHpfZcZnXOEpB3WPWHh1Qvbvv8lucICwuDk5OT6ebj4/OEiYmIyJL4O6sx9Wl3fNShmmnZz6fuYfTmWzgcmw4zOruFislsepZcXV2hVCrz9CIlJCTk6T3Kkd8w7AkJCVCpVKhaNf85gEJDQzFhwgTT/ZSUFPj4+KBu3bo8Z4mIiExq1dLi6yPbAQBV7dSIS9Xj872JaFOzKib3rIt6nvzOkFPOkaHSYDbFklqtRnBwMHbs2IHnnnvOtHzHjh3o27dvvuuEhITgzz//zLVs+/btaN68eYHnK2k0Gmg0eWefVigUUCjMriOOiIjKyMPfCdvGPYXFB2KwZH8UDl69iz5zD+Dllr6Y0LUOXO3zfqdQ2SvN72yz+vafMGEClixZgh9//BEREREYP348YmJiMHLkSADZvUKDBw82tR85ciSio6MxYcIERERE4Mcff8TSpUsxceJEuV4CERFZIDuNCu8/Uxc7J3RAr0aeMApg5eEYdJqxBz+HX6+wE/RS8ZhNzxIAvPTSS7h79y4+++wzxMXFoWHDhvjrr7/g5+cHAIiLi8s15lJAQAD++usvjB8/HvPmzYOXlxe+++47vPDCC3K9BCIismA+LraYN7AZhoQk4fPNF3D2ZjI++eM8Npy8ibDng0pl6hQqf2Y1zpIcOM4SERHlx2AwYOfOnQCAzp075xnF22gU+PVIDL7eehEPsvRQKSS8+VQNvNOlNjQqjvhd1krz+5vFUhFYLBER0ZOIT87Ep5vOY9v57AuO6nk6Ys7LTVDH3UHmZJatUs8NR0REZE48nKyxcFAwFv4vGFXt1IiIS0Gf7/dj+YEoDjNgJlgsERERlYAQAvfv38f9+/eLVfQ809ADW8e1R4c61ZClN+LTPy/gjZ+OITlDVw5p6UmwWCIiIioBnU6HOXPmYM6cOdDpilfwuDlYY/mwFpj6bAOoVQrsvJiAfvMO4PLt1DJOS0+CxRIREVE5kiQJQ9r4Y/2oNvCuYoOoO2noN+8Atp3Lf+oukh+LJSIiIhk09HbCprfbIqRGVaRpDRj163H8uD9K7liUDxZLREREMqlqr8Evr7fE4BA/CAF8tvkCwv6KgJGDWFYoLJaIiIhkpFIqMPXZBnj/mUAAwA//XsPE309z1O8KhMUSERGRzCRJwuiOtTCzf2MoFRLWn7iJd9ecYsFUQbBYIiIiqiBeDK6Oua80hUohYeOpWyyYKgizmhuOiIioolAoFGjevLnp59LSI8gTcyXg7ZUnsfHULSgVCszs3wiSJJXaNujxcLqTInC6EyIiksO2c/F4a+UJGIwCozrWxKRn6sodyaxwuhMiIiIL90xDD4Q9HwQAWLDnKpYd4LACcmGxREREVAJCCKSlpSEtLa3M5ngb0NwH73XPvkrus80XsPVsXJlshwrHYomIiKgEdDodZs6ciZkzZxZ7upOSGN2xJob8Nw7ThDWnceFWSplti/LHYomIiKgCkyQJH/euj/a1XZGhM2D4z8eQlKaVO1alwmKJiIioglMpFZj7SjP4V7XFzfsZGLXiOHQGo9yxKg0WS0RERGbAydYKiwc3h51aicNRSZj5d6TckSoNFktERERmora7A2YNaAwge1qU3ZEJMieqHFgsERERmZFnGnpiSIgfAODdNadxOyVT5kSWj8USERGRmQntWQ/1PR2RlKbFO7+d5JQoZYzFEhERUQkoFAo0btwYjRs3LtXpTorD2kqJuQObwlatxKFrSVi6/1q5br+y4XQnReB0J0REVFGtPhqDSevOQq1SYMuYdqjt7iB3pAqD050QERERBjT3QafAatDqjZi49jT0HE6gTLBYIiIiKgEhBLRaLbRabZlNd1IUSZLw1QuN4GitwukbyVi496osOSwdiyUiIqIS0Ol0CAsLQ1hYWJlOd1IUd0drTO3bAAAwZ+dlRMRxOpTSxmKJiIjIzPVr4o1u9d2hMwh8uOEsjLw6rlSxWCIiIjJzkiThs74NYa9R4WTMfaw8EiN3JIvCYomIiMgCeDhZY2K3OgCAr7ddREIqB6ssLSyWiIiILMSgEH80qu6E1Ew9Pt8cIXcci8FiiYiIyEIoFRKmPRcEhQT8efoW9l5KlDuSRWCxREREZEEaejthaJsAAMDUTeeh1XPspSfFYomIiKgEFAoF6tevj/r165f7dCdFGde1Nlzt1bh2Jw0/h1+XO47Zq1i/XSIiIjOhUqnQv39/9O/fHyqVSu44uThaW+H97nUBAHP+uYzE1CyZE5k3FktEREQW6MXg6gjydkJqlh6ztkfKHcessVgiIiKyQAqFhE+frQ8AWH0sFmdvJMucyHyxWCIiIioBrVaLqVOnYurUqdBqtXLHyVewnwv6NvGCEMDUP8/LNoeduWOxREREZME+6FEXNlZKHIu+hz/PxMkdxyyxWCIiIrJgnk42GN2xJgBg+raLyNIbZE5kflgsERERWbg32teAu6MGN+5l4JfwaLnjmB0WS0RERBbORq3Eu10DAQDf77qC5HSdzInMC4slIiKiSuCF4Oqo426P5Awd5u+5Inccs8JiiYiIqBJQKiSE9qgHAFh28Dpu3EuXOZH5YLFERERUAgqFArVr10bt2rUr3HQnBekYWA1talaFVm/ErO2X5I5jNiTBQRcKlZKSAicnJyQnJ8PR0VHuOERERE/k7I1k9Jm7HwCweUw7NPR2kjlR2SjN72/zKIWJiIioVARVd0LfJl4AgGl/RXCgymIwm2Lp3r17GDRoEJycnODk5IRBgwbh/v37ha4zdOhQSJKU69a6devyCUxERFRBTewWCLVSgYNX72LvpUS541R4ZlMsDRw4EKdOncK2bduwbds2nDp1CoMGDSpyvWeeeQZxcXGm219//VUOaYmIyNJptVpMmzYN06ZNq7DTnRTEx8UWQ9r4AQDC/roIg5G9S4VRyR2gOCIiIrBt2zYcOnQIrVq1AgAsXrwYISEhiIyMRGBgYIHrajQaeHh4lFdUIiKqRHQ68x2v6K1OtbDm2A1E3k7FuhM3MKC5j9yRKiyz6FkKDw+Hk5OTqVACgNatW8PJyQkHDx4sdN09e/bAzc0NderUwfDhw5GQkFDWcYmIiCq8KrZqvN2pFgDgm+2XkKHlNCgFMYtiKT4+Hm5ubnmWu7m5IT4+vsD1evTogV9//RW7du3CrFmzcPToUTz99NPIysoqcJ2srCykpKTkuhEREVmiQSF+8K5ig/iUTPx4IEruOBWWrMXSp59+mucE7Edvx44dAwBIkpRnfSFEvstzvPTSS+jVqxcaNmyIPn36YOvWrbh06RK2bNlS4DphYWGmk8idnJzg48NuSSIiskzWVkq81z37VJYFe67i7oOCOxMqM1nPWXr77bfx8ssvF9rG398fZ86cwe3bt/M8lpiYCHd392Jvz9PTE35+frh8+XKBbUJDQzFhwgTT/ZSUFBZMRERksZ5t7IXF+67h/K0UfL/rCj59toHckSocWYslV1dXuLq6FtkuJCQEycnJOHLkCFq2bAkAOHz4MJKTk9GmTZtib+/u3buIjY2Fp6dngW00Gg00Gk2xn5OIiMicKRQSPuxZD68uOYwVh6IxtI0//F3t5I5VoZjFOUv16tXDM888g+HDh+PQoUM4dOgQhg8fjt69e+e6Eq5u3brYsGEDAODBgweYOHEiwsPDcf36dezZswd9+vSBq6srnnvuObleChERWQhJkuDn5wc/P79CTwkxB21ruaJDnWrQGwVm/B0pd5wKxyyKJQD49ddfERQUhG7duqFbt25o1KgRfvnll1xtIiMjkZycDABQKpU4e/Ys+vbtizp16mDIkCGoU6cOwsPD4eDgIMdLICIiC2JlZYWhQ4di6NChsLKykjvOEwvtWReSBGw5G4eTMffkjlOhcG64InBuOCIiqizeW3saa4/fQEt/F6we0dqse8w4NxwRERGVugnd6kCjUuDI9ST8E8FxCXOwWCIiIioBrVaLGTNmYMaMGWY33UlBPJ1s8Hq7AADAV1sjoDcYZU5UMbBYIiIiKqH09HSkp6fLHaNUjexYE862VriamIY1x27IHadCYLFEREREJo7WVhjbuTYA4Jsdl5CWpZc5kfxYLBEREVEur7byg19VW9x5kIXF+67JHUd2LJaIiIgoF7VKgfe71wUALPr3GhJSM2VOJC8WS0RERJRHzyAPNPGpgnStAdO3Ve6BKlksERERUR6SJGFKn/oAgN+P38CJSjxQJYslIiKiEpAkCV5eXvDy8jLrwRsL09TXGS8GVwcAfLrpPIzGyjmONUfwLgJH8CYiososMTULT8/cg9QsPb5+IQgvtfCVO1KxcARvIiIiKhfVHDR4p0v2UALTt0UiOUMnc6Lyx2KJiIiICjWkjT9qudnjbpoW3+64JHeccsdiiYiIqAR0Oh1mz56N2bNnQ6ez7N4WK6UCn/ZpAAD45VA0LtxKkTlR+WKxREREVAJCCCQnJyM5ORmV4fTfdrVd0TPIAwajQOj6MzBUopO9WSwRERFRsXzapwEcrFU4fSMZyw9elztOuWGxRERERMXi5miN0B71AACztkfixj3LmkS4ICyWiIiIqNhebuGDlv4uSNca8NHGc5XiECSLJSIiIio2hULCtOeDoFYqsCcyERtO3pQ7UpljsURERESPpZabPcZ2rgUAmLLpPG7ez5A5UdlisURERFQCkiShWrVqqFatmsVOd1KYkR1qoqlvFaRm6vHumlMWPRUKpzspAqc7ISIiyt/1O2noMWcfMnQGfNSrHt5oX0PuSCac7oSIiIhk5+9qh4971weQPRXKxXjLHKySxRIRERGV2CstffB0XTdoDUa89esJPMjSyx2p1LFYIiIiKgGdTof58+dj/vz5Fj/dSWEkScL0FxvBw9EaVxPTELr+rMUNJ8BiiYiIqASEEEhMTERiYqLFFQePy9Veg3mvNoVKIeHP07fwc3i03JFKFYslIiIiemLBfi4I7Zk9uvcXWy7geHSSzIlKD4slIiIiKhWvtfVHryBP6AwCw38+jpi7ljEdCoslIiIiKhWSJGFG/0YI8nZCUpoWw5YfQXK6+Z/PxWKJiIiISo2tWoUlQ5rD0yn7hO9Rvx6HVm+UO9YTYbFEREREpcrd0RpLh7SAnVqJg1fvYuyqk9AbzLdgYrFERERUApIkwcnJCU5OTpVyupOi1PdyxMJBwVArFdh2Ph7vrj0Ng5lOicLpTorA6U6IiIhK7p8LtzFyxXHojQIDmldH2PONoFSUfXHJ6U6IiIjILHSp7445LzeFQgLWHLuBd347aXbnMLFYIiIiojLVq5EnvnulKayUEjaficPrPx1FmhlNi8JiiYiIqAR0Oh0WL16MxYsXV+rpToqrdyMv/Di0BWzVSuy7fAcvLzqEW/cz5I5VLCyWiIiISkAIgVu3buHWrVuVfrqT4mpfuxp+faMVnG2tcPZmMp6dux+Hr92VO1aRWCwRERFRuWnq64xNb7dDPU9H3HmgxatLDmPJvmswVuAr5VgsERERUbnycbHF+lFt0KexF/RGgS+2ROB/Sw9X2MNyLJaIiIio3Nmolfju5Sb4vF9D2FhlD17Zffa/+O1ITIXrZWKxRERERLKQJAmDWvthy9h2aOxTBamZenyw/iyeX3AQZ28kyx3PhMUSERERyapGNXusGxmCj3rVg71GhVOx9/HsvP0Y99tJRN1JkzseiyUiIqKSsrW1ha2trdwxLIJKqcAb7Wtg17sd0K+JF4QANp66hS7f7MXEtacRGZ8qWzZOd1IETndCRERU/s7dTMa3Oy5h58UE07K2tapiaJsAdAqsBpWy8P6e0vz+ZrFUBBZLRERE8jkRcw+L/72Gv8/HI+e8b1d7DZ5t7IXnm3mjgZdjvhMZs1gqRyyWiIiI5HfjXjp+CY/GmmOxuJf+/yOmV3e2Qee6bni6njtaBbjA2koJgMVSuWKxRERE+dHpdPj1118BAK+++iqsrKxkTlQ56AxG7I1MxIZTN/HPhdvIemhSXo1KgSY+VdAywAX1XK3QK7hmqXx/q540dHn58ssvsWXLFpw6dQpqtRr3798vch0hBKZOnYpFixbh3r17aNWqFebNm4cGDRqUfWAiIrJoQghER0ebfqbyYaVUoEt9d3Sp7450rR4HrtzFrosJ2HXxNm6nZOFwVBIORyXBmJVeats0m2JJq9Wif//+CAkJwdKlS4u1zvTp0/HNN99g+fLlqFOnDr744gt07doVkZGRcHBwKOPEREREVJZs1Sp0re+OrvXdIURDXLuThqNRSThyPQnhF2MRW0rbMZtiaerUqQCA5cuXF6u9EAKzZ8/G5MmT8fzzzwMAfvrpJ7i7u2PlypUYMWJEWUUlIiKiciZJEmpWs0fNavZ4uaUvUlJqwGlK6Ty3xY6zFBUVhfj4eHTr1s20TKPRoEOHDjh48GCB62VlZSElJSXXjYiIiCoviy2W4uPjAQDu7u65lru7u5sey09YWBicnJxMNx8fnzLNSURERBWbrMXSp59+CkmSCr0dO3bsibbx6NgLQoh8x2PIERoaiuTkZNMtNra0jngSERGROZL1nKW3334bL7/8cqFt/P39S/TcHh4eALJ7mDw9PU3LExIS8vQ2PUyj0UCj0ZRom0REVLlwuIDKQdZiydXVFa6urmXy3AEBAfDw8MCOHTvQtGlTANlX1O3duxdff/11mWyTiIgqD7VajQ8//FDuGFQOzOacpZiYGJw6dQoxMTEwGAw4deoUTp06hQcPHpja1K1bFxs2bACQffht3LhxmDZtGjZs2IBz585h6NChsLW1xcCBA+V6GURERGRmzGbogE8++QQ//fST6X5Ob9Hu3bvRsWNHAEBkZCSSk5NNbd5//31kZGRg9OjRpkEpt2/fzjGWiIiIqNg43UkRON0JERHlR6/XY82aNQCAAQMGQKUym/6HSqE0v7/5myUiIioBo9GIy5cvm34my2U25ywRERERyYHFEhEREVEhWCwRERERFYLFEhEREVEhWCwRERERFYJXwxUhZ2SFlJQUmZMQEVFFotVqkZmZCSD7O0KtVsuciB6W871dGiMkcZylIly7dg01a9aUOwYRERGVwNWrV1GjRo0neg72LBXBxcUFQPZ0K05OTjKnqVhSUlLg4+OD2NhYDtj5CO6bgnHf5I/7pWDcNwXjvilYcnIyfH19Td/jT4LFUhEUiuzTupycnPhGLICjoyP3TQG4bwrGfZM/7peCcd8UjPumYDnf40/0HKWQg4iIiMhisVgiIiIiKgSLpSJoNBpMmTIFGo1G7igVDvdNwbhvCsZ9kz/ul4Jx3xSM+6ZgpblveDUcERERUSHYs0RERERUCBZLRERERIVgsURERERUCBZLRERERIVgsVSI+fPnIyAgANbW1ggODsa+ffvkjiS7sLAwtGjRAg4ODnBzc0O/fv0QGRkpd6wKKSwsDJIkYdy4cXJHqRBu3ryJ//3vf6hatSpsbW3RpEkTHD9+XO5YstPr9fjoo48QEBAAGxsb1KhRA5999hmMRqPc0crdv//+iz59+sDLywuSJGHjxo25HhdC4NNPP4WXlxdsbGzQsWNHnD9/Xp6w5aywfaPT6TBp0iQEBQXBzs4OXl5eGDx4MG7duiVf4HJU1PvmYSNGjIAkSZg9e/ZjbYPFUgFWr16NcePGYfLkyTh58iTat2+PHj16ICYmRu5ostq7dy/eeustHDp0CDt27IBer0e3bt2QlpYmd7QK5ejRo1i0aBEaNWokd5QK4d69e2jbti2srKywdetWXLhwAbNmzUKVKlXkjia7r7/+GgsXLsTcuXMRERGB6dOnY8aMGfj+++/ljlbu0tLS0LhxY8ydOzffx6dPn45vvvkGc+fOxdGjR+Hh4YGuXbsiNTW1nJOWv8L2TXp6Ok6cOIGPP/4YJ06cwPr163Hp0iU8++yzMiQtf0W9b3Js3LgRhw8fhpeX1+NvRFC+WrZsKUaOHJlrWd26dcUHH3wgU6KKKSEhQQAQe/fulTtKhZGamipq164tduzYITp06CDeeecduSPJbtKkSaJdu3Zyx6iQevXqJV577bVcy55//nnxv//9T6ZEFQMAsWHDBtN9o9EoPDw8xFdffWValpmZKZycnMTChQtlSCifR/dNfo4cOSIAiOjo6PIJVUEUtG9u3LghvL29xblz54Sfn5/49ttvH+t52bOUD61Wi+PHj6Nbt265lnfr1g0HDx6UKVXFlJycDAClMlGhpXjrrbfQq1cvdOnSRe4oFcamTZvQvHlz9O/fH25ubmjatCkWL14sd6wKoV27dti5cycuXboEADh9+jT279+Pnj17ypysYomKikJ8fHyuz2WNRoMOHTrwczkfycnJkCSJvbcAjEYjBg0ahPfeew8NGjQo0XNwIt183LlzBwaDAe7u7rmWu7u7Iz4+XqZUFY8QAhMmTEC7du3QsGFDueNUCL/99huOHz+OY8eOyR2lQrl27RoWLFiACRMm4MMPP8SRI0cwduxYaDQaDB48WO54spo0aRKSk5NRt25dKJVKGAwGfPnll3jllVfkjlah5Hz25ve5HB0dLUekCiszMxMffPABBg4cyMl1kX2oW6VSYezYsSV+DhZLhZAkKdd9IUSeZZXZ22+/jTNnzmD//v1yR6kQYmNj8c4772D79u2wtraWO06FYjQa0bx5c0ybNg0A0LRpU5w/fx4LFiyo9MXS6tWrsWLFCqxcuRINGjTAqVOnMG7cOHh5eWHIkCFyx6tw+LlcOJ1Oh5dffhlGoxHz58+XO47sjh8/jjlz5uDEiRNP9D7hYbh8uLq6QqlU5ulFSkhIyPO/mspqzJgx2LRpE3bv3o3q1avLHadCOH78OBISEhAcHAyVSgWVSoW9e/fiu+++g0qlgsFgkDuibDw9PVG/fv1cy+rVq1fpL5gAgPfeew8ffPABXn75ZQQFBWHQoEEYP348wsLC5I5WoXh4eAAAP5cLodPpMGDAAERFRWHHjh3sVQKwb98+JCQkwNfX1/S5HB0djXfffRf+/v7Ffh4WS/lQq9UIDg7Gjh07ci3fsWMH2rRpI1OqikEIgbfffhvr16/Hrl27EBAQIHekCqNz5844e/YsTp06Zbo1b94cr776Kk6dOgWlUil3RNm0bds2zxATly5dgp+fn0yJKo709HQoFLk/ipVKZaUcOqAwAQEB8PDwyPW5rNVqsXfv3kr/uQz8f6F0+fJl/PPPP6hatarckSqEQYMG4cyZM7k+l728vPDee+/h77//Lvbz8DBcASZMmIBBgwahefPmCAkJwaJFixATE4ORI0fKHU1Wb731FlauXIk//vgDDg4Opv/lOTk5wcbGRuZ08nJwcMhz7padnR2qVq1a6c/pGj9+PNq0aYNp06ZhwIABOHLkCBYtWoRFixbJHU12ffr0wZdffglfX180aNAAJ0+exDfffIPXXntN7mjl7sGDB7hy5YrpflRUFE6dOgUXFxf4+vpi3LhxmDZtGmrXro3atWtj2rRpsLW1xcCBA2VMXT4K2zdeXl548cUXceLECWzevBkGg8H02ezi4gK1Wi1X7HJR1Pvm0cLRysoKHh4eCAwMLP5GnvxCPcs1b9484efnJ9RqtWjWrBkvjxfZl2Xmd1u2bJnc0SokDh3w//7880/RsGFDodFoRN26dcWiRYvkjlQhpKSkiHfeeUf4+voKa2trUaNGDTF58mSRlZUld7Ryt3v37nw/X4YMGSKEyB4+YMqUKcLDw0NoNBrx1FNPibNnz8obupwUtm+ioqIK/GzevXu33NHLXFHvm0eVZOgASQghHquEIyIiIqpEeM4SERERUSFYLBEREREVgsUSERERUSFYLBEREREVgsUSERERUSFYLBEREREVgsUSERERUSFYLBEREREVgsUSEVVqHTt2xLhx4/IsX758OapUqVLueYio4mGxRERERFQITqRLRBatY8eOpomMV6xYAaVSiVGjRuHzzz+HJEkypyMic8CeJSKyeD/99BNUKhUOHz6M7777Dt9++y2WLFkidywiMhOcSJeILFrHjh2RkJCA8+fPm3qSPvjgA2zatAkXLlxAx44dcfDgQajV6lzr6fV6WFtb4/79+zKkJqKKhD1LRGTxWrduneuQW0hICC5fvgyDwQAAePXVV3Hq1Klct88++0yuuERUwfCcJSKq9JycnFCrVq1cy9zc3GRKQ0QVDXuWiMjiHTp0KM/92rVrQ6lUypSIiMwJiyUisnixsbGYMGECIiMjsWrVKnz//fd455135I5FRGaCh+GIyOINHjwYGRkZaNmyJZRKJcaMGYM333xT7lhEZCZ4NRwRWbSOHTuiSZMmmD17ttxRiMhM8TAcERERUSFYLBEREREVgofhiIiIiArBniUiIiKiQrBYIiIiIioEiyUiIiKiQrBYIiIiIioEiyUiIiKiQrBYIiIiIioEiyUiIiKiQrBYIiIiIioEiyUiIiKiQvwf9+/Jquy6cvYAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Generating an array of x-coordinates, pH.\n",
    "pH_x = np.arange(0, 14.01, 0.01)\n",
    "\n",
    "# Generating an array of y-coordinates, charge.\n",
    "charge_y = estimateCharge('ARN', pH_x)\n",
    "\n",
    "# Finding the pH value which gives the charge closest to zero.\n",
    "pI = pH_x[np.argmin(abs(charge_y))]\n",
    "\n",
    "# Plot a graph of charge vs pH.\n",
    "plt.plot(pH_x, charge_y)\n",
    "plt.xlabel('pH')\n",
    "plt.ylabel('charge')\n",
    "plt.axhline(0, zorder = -1, color = 'lightgray')\n",
    "plt.axvline(pI, zorder = 2, color = 'gray', linestyle = '--')\n",
    "plt.text(pI + 0.2, 0.05, 'pI = {:.2f}'.format(pI))\n",
    "plt.title('Charge of Sequence ARN vs pH of Solution')\n",
    "plt.xlim(0, 14)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c4c50fc",
   "metadata": {},
   "source": [
    "The general idea of the estimateIsoelectric function shown below is to estimate the charge of the protein at various pH values until we encounter one where the magnitude of the net charge is less than the threshold value, delta. The threshold value is $1 \\times 10^{-5}$ by default, but can be adjusted depending on the error tolerance. \n",
    "\n",
    "The key feature differentiating the estimateIsoelectric function from the exhaustive method is the manner in which the pH values are chosen. If the magnitude of the charge of the protein at a tested pH value is greater than the threshold, the function needs to find a new pH value to be tested. If the starting test pH is 0.0, based on the graph above, we can see that the new test pH should be an increment of the old test pH to bring the charge closer to 0. Since the average proteome isoelectric point of most organisms is around 7 (source: [Proteome Isoelecric Point Database](http://isoelectricpointdb2.mimuw.edu.pl/index.html)), we fix the increment to be 7.0 by default, but adjustable depending on the proteome tested. If the new test pH is a better guess of pI (i.e. the pH gives a magnitude of charge which is closer to 0), we keep the same increment step when searching for the new test pH. \n",
    "\n",
    "If the new test pH is a worse guess, by inspecting the graph above, we know that our new test pH have overshot the pI value. Hence, we reduce the increment step by half to get our next test pH. The graph also implies that the increment should be negative when the charge is negative to bring the charge closer to 0.\n",
    "\n",
    "Admittedly, the scale of this project, and others in SP2273, is small enough such that there're not much added benefits in favour of the optimisation algorithim compared to the exhaustive method. Nonetheless, coming up with intelligent optimisation is a good habit to cultivate, especially when you deal with larger scale problems in your computing journey beyond SP2273. You may refer to the appendix for a numerical comparison between the optimisation algorithm and the exhaustive method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7f34c0fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "def estimateIsoelectric(chargeFunction, sequence, increment = 7.0, threshold = 1e-5):\n",
    "    \"\"\"Estimate the pI of a protein sequence.\"\"\"\n",
    "    \n",
    "    # In case the sequence is not a string.\n",
    "    sequence = str(sequence)\n",
    "    \n",
    "    # Starting test pH value, and the function used to compute protein charge.\n",
    "    best_pI = 0.0\n",
    "    minCharge = chargeFunction(sequence, best_pI)\n",
    "    \n",
    "    # The loop only stops when the charge of the protein is less than the threshold value.\n",
    "    while abs(minCharge) > threshold:\n",
    "        \n",
    "        # Compute the new test pH, and the protein charge of the new test pH.\n",
    "        pHtest = best_pI + increment\n",
    "        charge = chargeFunction(sequence, pHtest)\n",
    "        \n",
    "        # Steps if the new test pH is a better guess than the old test pH.\n",
    "        if abs(charge) < abs(minCharge):\n",
    "            \n",
    "            # Update protein charge.\n",
    "            minCharge = charge\n",
    "            \n",
    "            # Update the best guess pI value.\n",
    "            best_pI = pHtest\n",
    "        \n",
    "        # Steps if the new test pH is a worse guess than the old test pH.\n",
    "        else:\n",
    "            \n",
    "            # Half the increment step to prevent overshooting.\n",
    "            increment = abs(increment) / 2.0\n",
    "            \n",
    "            # Reverse the direction of increment if the protein charge is negative.\n",
    "            if minCharge < 0.0:\n",
    "                increment *= -1\n",
    "                \n",
    "    return best_pI"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5c3591d",
   "metadata": {},
   "source": [
    "### 3.3 Testing Algorithim on Amino Acids"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b46e7504",
   "metadata": {},
   "source": [
    "Before testing our algorithm against proteome database, we first start testing on individual amino acids."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3653e545",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pKa1</th>\n",
       "      <th>pKa2</th>\n",
       "      <th>pKa3</th>\n",
       "      <th>pI</th>\n",
       "      <th>calculated pI</th>\n",
       "      <th>percentage difference</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Alanine</th>\n",
       "      <td>2.34</td>\n",
       "      <td>9.69</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.00</td>\n",
       "      <td>6.02</td>\n",
       "      <td>-0.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Arginine</th>\n",
       "      <td>2.17</td>\n",
       "      <td>9.04</td>\n",
       "      <td>12.48</td>\n",
       "      <td>10.76</td>\n",
       "      <td>10.76</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Asparagine</th>\n",
       "      <td>2.02</td>\n",
       "      <td>8.80</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.41</td>\n",
       "      <td>5.41</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Aspartic acid</th>\n",
       "      <td>1.88</td>\n",
       "      <td>9.60</td>\n",
       "      <td>3.65</td>\n",
       "      <td>2.77</td>\n",
       "      <td>2.77</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cysteine</th>\n",
       "      <td>1.96</td>\n",
       "      <td>10.28</td>\n",
       "      <td>8.18</td>\n",
       "      <td>5.07</td>\n",
       "      <td>5.07</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Glutamic acid</th>\n",
       "      <td>2.19</td>\n",
       "      <td>9.67</td>\n",
       "      <td>4.25</td>\n",
       "      <td>3.22</td>\n",
       "      <td>3.22</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Glutamine</th>\n",
       "      <td>2.17</td>\n",
       "      <td>9.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.65</td>\n",
       "      <td>5.65</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Glycine</th>\n",
       "      <td>2.34</td>\n",
       "      <td>9.60</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.97</td>\n",
       "      <td>5.96</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Histidine</th>\n",
       "      <td>1.82</td>\n",
       "      <td>9.17</td>\n",
       "      <td>6.00</td>\n",
       "      <td>7.59</td>\n",
       "      <td>7.59</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hydroxyproline</th>\n",
       "      <td>1.82</td>\n",
       "      <td>9.65</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.74</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Isoleucine</th>\n",
       "      <td>2.36</td>\n",
       "      <td>9.60</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.02</td>\n",
       "      <td>5.99</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Leucine</th>\n",
       "      <td>2.36</td>\n",
       "      <td>9.60</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.98</td>\n",
       "      <td>5.99</td>\n",
       "      <td>-0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lysine</th>\n",
       "      <td>2.18</td>\n",
       "      <td>8.95</td>\n",
       "      <td>10.53</td>\n",
       "      <td>9.74</td>\n",
       "      <td>9.74</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Methionine</th>\n",
       "      <td>2.28</td>\n",
       "      <td>9.21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.74</td>\n",
       "      <td>5.74</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Phenylalanine</th>\n",
       "      <td>1.83</td>\n",
       "      <td>9.13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.48</td>\n",
       "      <td>5.48</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Proline</th>\n",
       "      <td>1.99</td>\n",
       "      <td>10.60</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.30</td>\n",
       "      <td>6.29</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Pyroglutamatic</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.68</td>\n",
       "      <td>0.00</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Serine</th>\n",
       "      <td>2.21</td>\n",
       "      <td>9.15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.68</td>\n",
       "      <td>5.67</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Threonine</th>\n",
       "      <td>2.09</td>\n",
       "      <td>9.10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.60</td>\n",
       "      <td>5.59</td>\n",
       "      <td>0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tryptophan</th>\n",
       "      <td>2.83</td>\n",
       "      <td>9.39</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.89</td>\n",
       "      <td>6.11</td>\n",
       "      <td>-3.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tyrosine</th>\n",
       "      <td>2.20</td>\n",
       "      <td>9.11</td>\n",
       "      <td>10.07</td>\n",
       "      <td>5.66</td>\n",
       "      <td>5.63</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Valine</th>\n",
       "      <td>2.32</td>\n",
       "      <td>9.62</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.96</td>\n",
       "      <td>5.96</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                pKa1   pKa2   pKa3     pI  calculated pI  \\\n",
       "Name                                                       \n",
       "Alanine         2.34   9.69    NaN   6.00           6.02   \n",
       "Arginine        2.17   9.04  12.48  10.76          10.76   \n",
       "Asparagine      2.02   8.80    NaN   5.41           5.41   \n",
       "Aspartic acid   1.88   9.60   3.65   2.77           2.77   \n",
       "Cysteine        1.96  10.28   8.18   5.07           5.07   \n",
       "Glutamic acid   2.19   9.67   4.25   3.22           3.22   \n",
       "Glutamine       2.17   9.13    NaN   5.65           5.65   \n",
       "Glycine         2.34   9.60    NaN   5.97           5.96   \n",
       "Histidine       1.82   9.17   6.00   7.59           7.59   \n",
       "Hydroxyproline  1.82   9.65    NaN    NaN           5.74   \n",
       "Isoleucine      2.36   9.60    NaN   6.02           5.99   \n",
       "Leucine         2.36   9.60    NaN   5.98           5.99   \n",
       "Lysine          2.18   8.95  10.53   9.74           9.74   \n",
       "Methionine      2.28   9.21    NaN   5.74           5.74   \n",
       "Phenylalanine   1.83   9.13    NaN   5.48           5.48   \n",
       "Proline         1.99  10.60    NaN   6.30           6.29   \n",
       "Pyroglutamatic   NaN    NaN    NaN   5.68           0.00   \n",
       "Serine          2.21   9.15    NaN   5.68           5.67   \n",
       "Threonine       2.09   9.10    NaN   5.60           5.59   \n",
       "Tryptophan      2.83   9.39    NaN   5.89           6.11   \n",
       "Tyrosine        2.20   9.11  10.07   5.66           5.63   \n",
       "Valine          2.32   9.62    NaN   5.96           5.96   \n",
       "\n",
       "                percentage difference  \n",
       "Name                                   \n",
       "Alanine                          -0.3  \n",
       "Arginine                          0.0  \n",
       "Asparagine                        0.0  \n",
       "Aspartic acid                     0.0  \n",
       "Cysteine                          0.0  \n",
       "Glutamic acid                     0.0  \n",
       "Glutamine                         0.0  \n",
       "Glycine                           0.2  \n",
       "Histidine                         0.0  \n",
       "Hydroxyproline                    NaN  \n",
       "Isoleucine                        0.5  \n",
       "Leucine                          -0.2  \n",
       "Lysine                            0.0  \n",
       "Methionine                        0.0  \n",
       "Phenylalanine                     0.0  \n",
       "Proline                           0.2  \n",
       "Pyroglutamatic                  100.0  \n",
       "Serine                            0.2  \n",
       "Threonine                         0.2  \n",
       "Tryptophan                       -3.7  \n",
       "Tyrosine                          0.5  \n",
       "Valine                            0.0  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add a new column to the amino acid dataframe, listing all the estimated pI values.\n",
    "predicted_pI = []\n",
    "for i in amino_acid['Symbol']:\n",
    "    predicted_pI.append(round(estimateIsoelectric(estimateCharge, i), 2))\n",
    "amino_acid['calculated pI'] = predicted_pI\n",
    "\n",
    "# Add a new column to the amino acid dataframe, listing the percentage difference between the estimated and actual pI values.\n",
    "amino_acid['percentage difference'] = round((amino_acid['pI'] - amino_acid['calculated pI']) / amino_acid['pI'] * 100, 1)\n",
    "\n",
    "# Load only columns of interest from the modified amino acid dataframe.\n",
    "amino_acid.loc[:, 'pKa1':]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c8b457b",
   "metadata": {},
   "source": [
    "We can see from the dataframe above that our algorithm performs admirably well. With the exception of the two amino acids with missing pK$_a$ or pI values, our algorithm managed to calculate the pI values of all amino acids to within 4% of the actual values. \n",
    "\n",
    "We are now ready to test our algorithm on proteome database."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "733989a8",
   "metadata": {},
   "source": [
    "### 3.4 Testing Algorithm on E. Coli Proteome"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d72e0376",
   "metadata": {},
   "source": [
    "The pI values of E.coli Proteome as obtained from the [Proteome Isoelecric Point Database](http://isoelectricpointdb2.mimuw.edu.pl/index.html) are calculated using 21 algorithms. We will compare our algorithms against the average of the 21 algorithms. We consider our algorithim to be accurate if our estimation lies within one standard deviation from the mean."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9b2930d7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>header</th>\n",
       "      <th>seq</th>\n",
       "      <th>mass</th>\n",
       "      <th>Bjellqvist</th>\n",
       "      <th>DTASelect</th>\n",
       "      <th>Dawson</th>\n",
       "      <th>EMBOSS</th>\n",
       "      <th>Grimsley</th>\n",
       "      <th>IPC2_peptide</th>\n",
       "      <th>IPC2_protein</th>\n",
       "      <th>...</th>\n",
       "      <th>Patrickios</th>\n",
       "      <th>ProMoST</th>\n",
       "      <th>Rodwell</th>\n",
       "      <th>Sillero</th>\n",
       "      <th>Solomon</th>\n",
       "      <th>Thurlkill</th>\n",
       "      <th>Toseland</th>\n",
       "      <th>Wikipedia</th>\n",
       "      <th>IPC2.protein.svr19</th>\n",
       "      <th>IPC2.peptide.svr19</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sp|P06960|OTC2_ECOLI Ornithine carbamoyltransf...</td>\n",
       "      <td>MATLKDIAIEAGVSLATVSRVLNDDPTLNVKEETKHRILEIAEKLE...</td>\n",
       "      <td>36210.35904</td>\n",
       "      <td>5.690</td>\n",
       "      <td>6.020</td>\n",
       "      <td>5.639</td>\n",
       "      <td>5.893</td>\n",
       "      <td>5.881</td>\n",
       "      <td>5.944</td>\n",
       "      <td>5.499</td>\n",
       "      <td>...</td>\n",
       "      <td>4.215</td>\n",
       "      <td>5.753</td>\n",
       "      <td>5.601</td>\n",
       "      <td>5.944</td>\n",
       "      <td>5.639</td>\n",
       "      <td>5.931</td>\n",
       "      <td>5.766</td>\n",
       "      <td>5.588</td>\n",
       "      <td>5.510</td>\n",
       "      <td>5.936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sp|P06993|MALT_ECOLI HTH-type transcriptional ...</td>\n",
       "      <td>MSDLYKKHFLKLLDFTPAQFTSLLTLAAQLKADKKNGKEVQKLTGK...</td>\n",
       "      <td>36827.14434</td>\n",
       "      <td>5.626</td>\n",
       "      <td>5.944</td>\n",
       "      <td>5.563</td>\n",
       "      <td>5.804</td>\n",
       "      <td>5.792</td>\n",
       "      <td>5.868</td>\n",
       "      <td>5.410</td>\n",
       "      <td>...</td>\n",
       "      <td>4.240</td>\n",
       "      <td>5.563</td>\n",
       "      <td>5.525</td>\n",
       "      <td>5.868</td>\n",
       "      <td>5.563</td>\n",
       "      <td>5.855</td>\n",
       "      <td>5.690</td>\n",
       "      <td>5.512</td>\n",
       "      <td>5.468</td>\n",
       "      <td>5.866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sp|P07639|AROB_ECOLI 3-dehydroquinate synthase...</td>\n",
       "      <td>MLIPSKLSRPVRLDHTVVRERLLAKLSGANNFRLALITSPAGYGKT...</td>\n",
       "      <td>103118.45694</td>\n",
       "      <td>6.033</td>\n",
       "      <td>6.491</td>\n",
       "      <td>6.020</td>\n",
       "      <td>6.465</td>\n",
       "      <td>6.440</td>\n",
       "      <td>6.415</td>\n",
       "      <td>5.677</td>\n",
       "      <td>...</td>\n",
       "      <td>4.279</td>\n",
       "      <td>6.109</td>\n",
       "      <td>6.008</td>\n",
       "      <td>6.402</td>\n",
       "      <td>6.020</td>\n",
       "      <td>6.504</td>\n",
       "      <td>6.249</td>\n",
       "      <td>6.033</td>\n",
       "      <td>5.821</td>\n",
       "      <td>6.299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sp|P08191|FIMH_ECOLI Type 1 fimbrin D-mannose ...</td>\n",
       "      <td>MERIVVTLGERSYPITIASGLFNEPASFLPLKSGEQVMLVTNETLA...</td>\n",
       "      <td>38880.98884</td>\n",
       "      <td>5.715</td>\n",
       "      <td>6.046</td>\n",
       "      <td>5.664</td>\n",
       "      <td>5.931</td>\n",
       "      <td>5.855</td>\n",
       "      <td>5.982</td>\n",
       "      <td>5.537</td>\n",
       "      <td>...</td>\n",
       "      <td>4.202</td>\n",
       "      <td>5.779</td>\n",
       "      <td>5.639</td>\n",
       "      <td>5.982</td>\n",
       "      <td>5.664</td>\n",
       "      <td>5.970</td>\n",
       "      <td>5.766</td>\n",
       "      <td>5.626</td>\n",
       "      <td>5.519</td>\n",
       "      <td>5.975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sp|P09831|GLTB_ECOLI Glutamate synthase [NADPH...</td>\n",
       "      <td>MKRVITLFAVLLMGWSVNAWSFACKTANGTAIPIGGGSANVYVNLA...</td>\n",
       "      <td>31473.48884</td>\n",
       "      <td>7.585</td>\n",
       "      <td>7.732</td>\n",
       "      <td>7.658</td>\n",
       "      <td>7.878</td>\n",
       "      <td>6.561</td>\n",
       "      <td>7.293</td>\n",
       "      <td>6.795</td>\n",
       "      <td>...</td>\n",
       "      <td>0.502</td>\n",
       "      <td>7.190</td>\n",
       "      <td>7.629</td>\n",
       "      <td>8.053</td>\n",
       "      <td>7.775</td>\n",
       "      <td>7.761</td>\n",
       "      <td>6.576</td>\n",
       "      <td>7.585</td>\n",
       "      <td>7.530</td>\n",
       "      <td>7.500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4445</th>\n",
       "      <td>sp|P15005-2|MCRB-2_ECOLI Isoform of P15005  Is...</td>\n",
       "      <td>MNIFKPISYIASLAPREVTLLALV</td>\n",
       "      <td>2660.25364</td>\n",
       "      <td>8.346</td>\n",
       "      <td>8.829</td>\n",
       "      <td>9.063</td>\n",
       "      <td>9.297</td>\n",
       "      <td>8.887</td>\n",
       "      <td>7.995</td>\n",
       "      <td>7.732</td>\n",
       "      <td>...</td>\n",
       "      <td>7.614</td>\n",
       "      <td>8.463</td>\n",
       "      <td>9.019</td>\n",
       "      <td>9.019</td>\n",
       "      <td>9.721</td>\n",
       "      <td>8.858</td>\n",
       "      <td>9.107</td>\n",
       "      <td>9.180</td>\n",
       "      <td>8.273</td>\n",
       "      <td>8.273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4446</th>\n",
       "      <td>sp|P63284-2|CLPB-2_ECOLI Isoform of P63284  Is...</td>\n",
       "      <td>MSKTESYCLEDALNDLFIPETTIETILKRLTIKKNIILQGPPGVGK...</td>\n",
       "      <td>34659.50584</td>\n",
       "      <td>5.258</td>\n",
       "      <td>5.359</td>\n",
       "      <td>5.130</td>\n",
       "      <td>5.016</td>\n",
       "      <td>4.991</td>\n",
       "      <td>5.308</td>\n",
       "      <td>5.245</td>\n",
       "      <td>...</td>\n",
       "      <td>4.037</td>\n",
       "      <td>5.270</td>\n",
       "      <td>5.041</td>\n",
       "      <td>5.321</td>\n",
       "      <td>5.118</td>\n",
       "      <td>5.067</td>\n",
       "      <td>5.080</td>\n",
       "      <td>4.978</td>\n",
       "      <td>5.123</td>\n",
       "      <td>5.123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4447</th>\n",
       "      <td>sp|P75960-2|NPD-2_ECOLI Isoform of P75960  Iso...</td>\n",
       "      <td>MNDQGAEDQRQALKKYTIDLTERAEQGKLDPVIGRDEEIRRTIQVL...</td>\n",
       "      <td>79870.23204</td>\n",
       "      <td>5.410</td>\n",
       "      <td>5.563</td>\n",
       "      <td>5.308</td>\n",
       "      <td>5.258</td>\n",
       "      <td>5.270</td>\n",
       "      <td>5.525</td>\n",
       "      <td>5.308</td>\n",
       "      <td>...</td>\n",
       "      <td>4.495</td>\n",
       "      <td>5.436</td>\n",
       "      <td>5.245</td>\n",
       "      <td>5.525</td>\n",
       "      <td>5.308</td>\n",
       "      <td>5.321</td>\n",
       "      <td>5.321</td>\n",
       "      <td>5.169</td>\n",
       "      <td>5.326</td>\n",
       "      <td>5.326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4448</th>\n",
       "      <td>sp|Q59385-2|COPA-2_ECOLI Isoform of Q59385  Is...</td>\n",
       "      <td>MEKPRVLVLTGAGISAESGIRTFRAADGLWEEHRVEDVATPEGFDR...</td>\n",
       "      <td>26716.38314</td>\n",
       "      <td>5.423</td>\n",
       "      <td>5.703</td>\n",
       "      <td>5.359</td>\n",
       "      <td>5.512</td>\n",
       "      <td>5.512</td>\n",
       "      <td>5.652</td>\n",
       "      <td>5.245</td>\n",
       "      <td>...</td>\n",
       "      <td>4.012</td>\n",
       "      <td>5.474</td>\n",
       "      <td>5.308</td>\n",
       "      <td>5.639</td>\n",
       "      <td>5.359</td>\n",
       "      <td>5.575</td>\n",
       "      <td>5.461</td>\n",
       "      <td>5.283</td>\n",
       "      <td>5.293</td>\n",
       "      <td>5.293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4449</th>\n",
       "      <td>sp|Q59385-2|COPA-2_ECOLI Isoform of Q59385  Is...</td>\n",
       "      <td>MSQTIDLTLDGLSCGHCVKRVKESLEQRPDVEQADVSITEAHVTGT...</td>\n",
       "      <td>7481.31954</td>\n",
       "      <td>4.902</td>\n",
       "      <td>5.029</td>\n",
       "      <td>4.762</td>\n",
       "      <td>4.673</td>\n",
       "      <td>4.622</td>\n",
       "      <td>4.940</td>\n",
       "      <td>4.952</td>\n",
       "      <td>...</td>\n",
       "      <td>3.897</td>\n",
       "      <td>4.914</td>\n",
       "      <td>4.673</td>\n",
       "      <td>4.952</td>\n",
       "      <td>4.762</td>\n",
       "      <td>4.711</td>\n",
       "      <td>4.698</td>\n",
       "      <td>4.647</td>\n",
       "      <td>4.862</td>\n",
       "      <td>4.912</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4450 rows  24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 header  \\\n",
       "0     sp|P06960|OTC2_ECOLI Ornithine carbamoyltransf...   \n",
       "1     sp|P06993|MALT_ECOLI HTH-type transcriptional ...   \n",
       "2     sp|P07639|AROB_ECOLI 3-dehydroquinate synthase...   \n",
       "3     sp|P08191|FIMH_ECOLI Type 1 fimbrin D-mannose ...   \n",
       "4     sp|P09831|GLTB_ECOLI Glutamate synthase [NADPH...   \n",
       "...                                                 ...   \n",
       "4445  sp|P15005-2|MCRB-2_ECOLI Isoform of P15005  Is...   \n",
       "4446  sp|P63284-2|CLPB-2_ECOLI Isoform of P63284  Is...   \n",
       "4447  sp|P75960-2|NPD-2_ECOLI Isoform of P75960  Iso...   \n",
       "4448  sp|Q59385-2|COPA-2_ECOLI Isoform of Q59385  Is...   \n",
       "4449  sp|Q59385-2|COPA-2_ECOLI Isoform of Q59385  Is...   \n",
       "\n",
       "                                                    seq          mass  \\\n",
       "0     MATLKDIAIEAGVSLATVSRVLNDDPTLNVKEETKHRILEIAEKLE...   36210.35904   \n",
       "1     MSDLYKKHFLKLLDFTPAQFTSLLTLAAQLKADKKNGKEVQKLTGK...   36827.14434   \n",
       "2     MLIPSKLSRPVRLDHTVVRERLLAKLSGANNFRLALITSPAGYGKT...  103118.45694   \n",
       "3     MERIVVTLGERSYPITIASGLFNEPASFLPLKSGEQVMLVTNETLA...   38880.98884   \n",
       "4     MKRVITLFAVLLMGWSVNAWSFACKTANGTAIPIGGGSANVYVNLA...   31473.48884   \n",
       "...                                                 ...           ...   \n",
       "4445                           MNIFKPISYIASLAPREVTLLALV    2660.25364   \n",
       "4446  MSKTESYCLEDALNDLFIPETTIETILKRLTIKKNIILQGPPGVGK...   34659.50584   \n",
       "4447  MNDQGAEDQRQALKKYTIDLTERAEQGKLDPVIGRDEEIRRTIQVL...   79870.23204   \n",
       "4448  MEKPRVLVLTGAGISAESGIRTFRAADGLWEEHRVEDVATPEGFDR...   26716.38314   \n",
       "4449  MSQTIDLTLDGLSCGHCVKRVKESLEQRPDVEQADVSITEAHVTGT...    7481.31954   \n",
       "\n",
       "      Bjellqvist  DTASelect  Dawson  EMBOSS  Grimsley  IPC2_peptide  \\\n",
       "0          5.690      6.020   5.639   5.893     5.881         5.944   \n",
       "1          5.626      5.944   5.563   5.804     5.792         5.868   \n",
       "2          6.033      6.491   6.020   6.465     6.440         6.415   \n",
       "3          5.715      6.046   5.664   5.931     5.855         5.982   \n",
       "4          7.585      7.732   7.658   7.878     6.561         7.293   \n",
       "...          ...        ...     ...     ...       ...           ...   \n",
       "4445       8.346      8.829   9.063   9.297     8.887         7.995   \n",
       "4446       5.258      5.359   5.130   5.016     4.991         5.308   \n",
       "4447       5.410      5.563   5.308   5.258     5.270         5.525   \n",
       "4448       5.423      5.703   5.359   5.512     5.512         5.652   \n",
       "4449       4.902      5.029   4.762   4.673     4.622         4.940   \n",
       "\n",
       "      IPC2_protein  ...  Patrickios  ProMoST  Rodwell  Sillero  Solomon  \\\n",
       "0            5.499  ...       4.215    5.753    5.601    5.944    5.639   \n",
       "1            5.410  ...       4.240    5.563    5.525    5.868    5.563   \n",
       "2            5.677  ...       4.279    6.109    6.008    6.402    6.020   \n",
       "3            5.537  ...       4.202    5.779    5.639    5.982    5.664   \n",
       "4            6.795  ...       0.502    7.190    7.629    8.053    7.775   \n",
       "...            ...  ...         ...      ...      ...      ...      ...   \n",
       "4445         7.732  ...       7.614    8.463    9.019    9.019    9.721   \n",
       "4446         5.245  ...       4.037    5.270    5.041    5.321    5.118   \n",
       "4447         5.308  ...       4.495    5.436    5.245    5.525    5.308   \n",
       "4448         5.245  ...       4.012    5.474    5.308    5.639    5.359   \n",
       "4449         4.952  ...       3.897    4.914    4.673    4.952    4.762   \n",
       "\n",
       "      Thurlkill  Toseland  Wikipedia  IPC2.protein.svr19  IPC2.peptide.svr19  \n",
       "0         5.931     5.766      5.588               5.510               5.936  \n",
       "1         5.855     5.690      5.512               5.468               5.866  \n",
       "2         6.504     6.249      6.033               5.821               6.299  \n",
       "3         5.970     5.766      5.626               5.519               5.975  \n",
       "4         7.761     6.576      7.585               7.530               7.500  \n",
       "...         ...       ...        ...                 ...                 ...  \n",
       "4445      8.858     9.107      9.180               8.273               8.273  \n",
       "4446      5.067     5.080      4.978               5.123               5.123  \n",
       "4447      5.321     5.321      5.169               5.326               5.326  \n",
       "4448      5.575     5.461      5.283               5.293               5.293  \n",
       "4449      4.711     4.698      4.647               4.862               4.912  \n",
       "\n",
       "[4450 rows x 24 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import the proteome database as dataframe, including only columns relevant to pI calculation.\n",
    "ecoli = pd.read_csv('ecoli_isoelectric.csv', usecols = range(24))\n",
    "ecoli"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4cd8266b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add new columns to the ecoli dataframe, showing the mean and standard deviation of pI values obtained from 21 algorithms. \n",
    "ecoli['average pI'] = [round(ecoli.iloc[i, 3:].mean(), 2) for i in range(ecoli.shape[0])]\n",
    "ecoli['standard deviation'] = [round(ecoli.iloc[i, 3:23].std(), 2) for i in range(ecoli.shape[0])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3c99a0e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def within_range(chargeFunction):\n",
    "    \"\"\"Check whether the estimated pI value is within one standard deviation from the mean for each sequence.\"\"\"\n",
    "    \n",
    "    # Add a new column to the ecoli dataframe, listing all the estimated pI values.\n",
    "    predicted_ecoli_pI = []\n",
    "    for i in ecoli['seq']:\n",
    "        predicted_ecoli_pI.append(round(estimateIsoelectric(chargeFunction, i), 2))\n",
    "    ecoli['calculated pI'] = predicted_ecoli_pI\n",
    "    \n",
    "    # Add a new column to the ecoli dataframe, listing whether the estimated pI value is within one sd from the mean.\n",
    "    ecoli['within range?'] = ecoli.apply(lambda x: \n",
    "                                   True if x['calculated pI'] >= x['average pI'] - x['standard deviation'] \n",
    "                                   and x['calculated pI'] <= x['average pI'] + x['standard deviation'] else False, axis = 1)\n",
    "    \n",
    "    # Show only the sequence where the estimated pI value lies more than one sd from the mean.\n",
    "    return ecoli[ecoli['within range?'] == False].loc[:, 'average pI':]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8b5fa11e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>average pI</th>\n",
       "      <th>standard deviation</th>\n",
       "      <th>calculated pI</th>\n",
       "      <th>within range?</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>4.70</td>\n",
       "      <td>0.14</td>\n",
       "      <td>4.54</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>271</th>\n",
       "      <td>10.77</td>\n",
       "      <td>0.51</td>\n",
       "      <td>11.29</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340</th>\n",
       "      <td>4.31</td>\n",
       "      <td>0.15</td>\n",
       "      <td>4.11</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365</th>\n",
       "      <td>10.68</td>\n",
       "      <td>0.49</td>\n",
       "      <td>11.18</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>416</th>\n",
       "      <td>4.51</td>\n",
       "      <td>0.14</td>\n",
       "      <td>4.35</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4421</th>\n",
       "      <td>8.97</td>\n",
       "      <td>0.61</td>\n",
       "      <td>9.64</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4427</th>\n",
       "      <td>4.34</td>\n",
       "      <td>0.28</td>\n",
       "      <td>3.96</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4428</th>\n",
       "      <td>8.79</td>\n",
       "      <td>0.61</td>\n",
       "      <td>9.54</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4433</th>\n",
       "      <td>8.95</td>\n",
       "      <td>0.62</td>\n",
       "      <td>9.64</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4445</th>\n",
       "      <td>8.79</td>\n",
       "      <td>0.61</td>\n",
       "      <td>9.54</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows  4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      average pI  standard deviation  calculated pI  within range?\n",
       "46          4.70                0.14           4.54          False\n",
       "271        10.77                0.51          11.29          False\n",
       "340         4.31                0.15           4.11          False\n",
       "365        10.68                0.49          11.18          False\n",
       "416         4.51                0.14           4.35          False\n",
       "...          ...                 ...            ...            ...\n",
       "4421        8.97                0.61           9.64          False\n",
       "4427        4.34                0.28           3.96          False\n",
       "4428        8.79                0.61           9.54          False\n",
       "4433        8.95                0.62           9.64          False\n",
       "4445        8.79                0.61           9.54          False\n",
       "\n",
       "[100 rows x 4 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "within_range(estimateCharge)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91621327",
   "metadata": {},
   "source": [
    "From here, we can see that despite ignoring the effect of changing pK$_a$ values, our algorithm performs surprisingly well. Out of 4450 proteins, only 100 of them has pI values lying outside the acceptable range.\n",
    "\n",
    "Next, we attempt to improve the accuracy of our algorithm by considering the effect of varying pK$_a$ values due to neighbouring amino acids. Our goal is to have less than 100 proteins with pI values lying outside the acceptable range."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0ece5e9",
   "metadata": {},
   "source": [
    "### 3.5 Consider the effect of varying dissociation constants of termini"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "429935b9",
   "metadata": {},
   "source": [
    "The R groups can be categorised into two groups: electron withdrawing group and electron donating group. Like the name suggests, an electron withdrawing group reduces the electron density of neighbouring groups. This causes deprotonated C-terminal to be more stable as the deprotonated C-terminal becomes less negatively charged. This causes the C-terminal to have a higher tendency to get deprotonated, making it a stronger acid, resulting in a lower pK$_a$ value. An electron withdrawing group also causes a protonated N-terminal to become less stable as the protonated N-terminal becomes more positively charged. This causes the protonated N-terminal to have a higher tendency to lose a proton, also resulting in a lower pK$_a$ value. An electron donating group has the opposite effect on both C-terminal and N-terminal.\n",
    "\n",
    "With this knowledge, we will be modifying the pK$_a$ values of the C- and N-termini of a protein depending on the type of R groups making up the sequence. All we need is to figure out whether each R group is electron withdrawing or electron donating. While it is technically possible to study each R group one-by-one manually, we shall make use of the amino acid data frame to help us decide quickly.\n",
    "\n",
    "The figure below shows the structural formula of a glycine.\n",
    "<img src='gly.gif' width='200'/>\n",
    "\n",
    "The figure shows that the R group of glycine is H, which is regarded as the standard and has no electron withdrawing or donating effect. Hence, we use the pK$_a$ value of the C-terminal of glycine, 2.34 as a gauge on whether the R group of an amino acid is electron withdrawing or donating. If the pK$_a$ value of the C-terminal of an amino acid is less than 2.34, it means that the deprotonated C-terminal is more stable, implying that the R group is an electron withdrawing group. If the pK$_a$ value is more than 2.34, the R group is an electron donating group. If the pK$_a$ value is exactly 2.34, the R group has no effect on neighbouring groups.\n",
    "\n",
    "With this, we build a dictionary to capture the electron donating and withdrawing property of each R group. For simplicity, we assume that all R groups will cause a change of 0.5 in the pK$_a$ value of its immediate neighbour. Since both the C- and N-termini experience the same direction of change in pk$_a$ value, we will also assume that each R group will cause the same change in both C- and N-terminal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7f0d5fb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "fixValue = []\n",
    "\n",
    "# Extract the pKa value of the C-terminal of an amino acid.\n",
    "for i in amino_acid['pKa1']:\n",
    "    \n",
    "    # Check if the R group is electron withdrawing.\n",
    "    if i < 2.34:\n",
    "        fixValue.append(-0.5)\n",
    "    \n",
    "    # Check if the R group is electron donating.\n",
    "    elif i > 2.34:\n",
    "        fixValue.append(0.5)\n",
    "    \n",
    "    # If the R group is neutral.\n",
    "    else:\n",
    "        fixValue.append(0)\n",
    "\n",
    "# Key: Amino acid, Value: assumed change in pKa value\n",
    "pKafixed = dict(zip(amino_acid['Symbol'], fixValue))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be519d3a",
   "metadata": {},
   "source": [
    "With the dictionary, we will now make a new function to estimate the charge of a protein at various pH values, estimateCharge2. For the most parts, this function is the same as estimateCharge, with the addition of new lines to amend the pK$_a$ values of the C- and N-terminal before calculating their respective charges. Since we expect the effect of each R group on a terminal to be weaker if the R group is further from the terminal, a simple weighing function is introduced when computing the change. The default change of 0.5 is divided by the distance of the amino acid from the terminal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "917c2b41",
   "metadata": {},
   "outputs": [],
   "source": [
    "def estimateCharge2(sequence, pH):\n",
    "    \"\"\"Use pKa values to estimate the charge of a sequence of amino acids at a given pH.\n",
    "    Take into account the effect of the immediate neighbour of the C-terminus.\"\"\"\n",
    "    total = 0.0\n",
    "    pKaN = pKaDict_NH2.get(sequence[0])\n",
    "    \n",
    "    # Consider the effect of all R groups starting from the second amino acid.\n",
    "    for i in range(1, len(sequence)):\n",
    "        \n",
    "        # Obtain the unweighted change in pKa value associated to the R group.\n",
    "        fixedN = pKafixed.get(sequence[i])\n",
    "        \n",
    "        # In case the amino acid is not among the ones listed in the amino_acid data frame.\n",
    "        if fixedN is not None:\n",
    "            \n",
    "            # Change the pKa value of the N-terminal, weighted by the distance from ther terminal.\n",
    "            pKaN += fixedN / i\n",
    "            \n",
    "    rN = 10.0 ** (pH - pKaN)\n",
    "    dissociatedN = rN / (rN + 1.0)\n",
    "    chargeN = 1.0 - dissociatedN\n",
    "    total += chargeN\n",
    "    for aminoAcid in sequence:\n",
    "        pKaR = pKaDict_R.get(aminoAcid)\n",
    "        if pKaR is not None:\n",
    "            rR = 10.0 ** (pH - pKaR)\n",
    "            dissociatedR = rR / (rR + 1.0)\n",
    "            if isAcid[aminoAcid]:\n",
    "                chargeR = -1.0 * dissociatedR\n",
    "            else:\n",
    "                chargeR = 1.0 - dissociatedR\n",
    "            total += chargeR\n",
    "    pKaC = pKaDict_COOH.get(sequence[-1])\n",
    "    \n",
    "    # Consider the effect of all R groups starting from the second amino acid counting from the right.\n",
    "    for i in range(-2, -1*len(sequence)-1, -1):\n",
    "        \n",
    "        # Obtain the unweighted change in pKa value associated to the R group.\n",
    "        fixedC = pKafixed.get(sequence[i])\n",
    "        \n",
    "        # In case the amino acid is not among the ones listed in the amino_acid data frame.\n",
    "        if fixedC is not None:\n",
    "            \n",
    "            # Change the pKa value of the C-terminal, weighted by the distance from ther terminal.\n",
    "            pKaC += fixedC / (i * (-1) - 1)\n",
    "    \n",
    "    rC = 10.0 ** (pH - pKaC)\n",
    "    dissociatedC = rC / (rC + 1.0)\n",
    "    chargeC = -1.0 * dissociatedC\n",
    "    total += chargeC\n",
    "    return total"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ff4c0f49",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>average pI</th>\n",
       "      <th>standard deviation</th>\n",
       "      <th>calculated pI</th>\n",
       "      <th>within range?</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>4.70</td>\n",
       "      <td>0.14</td>\n",
       "      <td>4.54</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340</th>\n",
       "      <td>4.31</td>\n",
       "      <td>0.15</td>\n",
       "      <td>4.11</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365</th>\n",
       "      <td>10.68</td>\n",
       "      <td>0.49</td>\n",
       "      <td>11.18</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>416</th>\n",
       "      <td>4.51</td>\n",
       "      <td>0.14</td>\n",
       "      <td>4.35</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>511</th>\n",
       "      <td>4.17</td>\n",
       "      <td>0.16</td>\n",
       "      <td>4.00</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4403</th>\n",
       "      <td>4.28</td>\n",
       "      <td>0.21</td>\n",
       "      <td>3.95</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4412</th>\n",
       "      <td>3.79</td>\n",
       "      <td>0.57</td>\n",
       "      <td>2.91</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4413</th>\n",
       "      <td>4.80</td>\n",
       "      <td>0.16</td>\n",
       "      <td>4.59</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4421</th>\n",
       "      <td>8.97</td>\n",
       "      <td>0.61</td>\n",
       "      <td>9.66</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4427</th>\n",
       "      <td>4.34</td>\n",
       "      <td>0.28</td>\n",
       "      <td>3.95</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>75 rows  4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      average pI  standard deviation  calculated pI  within range?\n",
       "46          4.70                0.14           4.54          False\n",
       "340         4.31                0.15           4.11          False\n",
       "365        10.68                0.49          11.18          False\n",
       "416         4.51                0.14           4.35          False\n",
       "511         4.17                0.16           4.00          False\n",
       "...          ...                 ...            ...            ...\n",
       "4403        4.28                0.21           3.95          False\n",
       "4412        3.79                0.57           2.91          False\n",
       "4413        4.80                0.16           4.59          False\n",
       "4421        8.97                0.61           9.66          False\n",
       "4427        4.34                0.28           3.95          False\n",
       "\n",
       "[75 rows x 4 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "within_range(estimateCharge2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9da77b2",
   "metadata": {},
   "source": [
    "Previously, without considering the effect of R groups, the pI values of 100 proteins out of 4450 is more than one sd from the mean. By using a simple algorithm to consider the effect of R groups on the pK$_a$ values of the C- and N-terminal, only 75 proteins fall outside the range. This shows that our simple algorithm does improve the estimation of the protein pI values."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88afe27e",
   "metadata": {},
   "source": [
    "## 4. Future Work"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c10f1c2",
   "metadata": {},
   "source": [
    "For those interested in picking up this project, the following are three possible avenues you can work on:\n",
    "1) Explore ways to more accurately reflect the changes in pK$_a$ values due to neighbouring amino acids.  \n",
    "2) Consider the effect of protein folding on the pK$_a$ values of component amino acids. This effect has been ignored throughout this notebook.  \n",
    "3) Study why the algorithm used in this notebook worked as well as it did. Like holy @#$%! I myself am damn surprised by how accurate this is. o_O"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc255c20",
   "metadata": {},
   "source": [
    "## 5. Appendix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ece2ff5",
   "metadata": {},
   "source": [
    "Here, we make a comparison between the time needed to execute to estimate the pI values using the exhaustive method and the optimisation algorithm. As usual, Python already has a module for that. We will use the 'ARN' sequence as an example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "58222c66",
   "metadata": {},
   "outputs": [],
   "source": [
    "from timeit import default_timer as timer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ff29a258",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The pI value of 11.09 is obtained in 1.29e-03 seconds\n"
     ]
    }
   ],
   "source": [
    "start = timer()\n",
    "pH_x = np.arange(0, 14.01, 0.01)\n",
    "charge_y = estimateCharge('ARN', pH_x)\n",
    "pI = pH_x[np.argmin(abs(charge_y))]\n",
    "end = timer()\n",
    "time_diff = end - start\n",
    "print('The pI value of {0:.2f} is obtained in {1:.2e} seconds'.format(pI, time_diff))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2cc4a3d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The pI value of 11.09 is obtained in 1.92e-04 seconds\n"
     ]
    }
   ],
   "source": [
    "start = timer()\n",
    "pI = estimateIsoelectric(estimateCharge, 'ARN')\n",
    "end = timer()\n",
    "time_diff = end - start\n",
    "print('The pI value of {0:.2f} is obtained in {1:.2e} seconds'.format(pI, time_diff))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fe450e0",
   "metadata": {},
   "source": [
    "Both methods give us the same pI value. However, the optimisation algorithm does so in approximately one-tenth of the time taken by the exhaustive method. The difference is not significant for a small scale project such as this. But once the project turns large scale, the time saved will become significant."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
